# Introduction to structural bioinformatics

Structural bioinformatics is a multidisciplinary area enriched by chemistry, physics, informatics and many others. Although, it could be focused on different biological macromolecules, here the emphasis will be focused on proteins.

One of the first protein structure elucidated was myoglobin and it triggered the study of the role of the structure of proteins and its biological functions


Identify a protein related to your study that could be further analyzed.

## Protein structures

Difference between the levels of protein structure
primary structure is the basic linear representation of aminoacids. Natural aminoacids and modified or rare aminoacids display physicochemical rich information and could be represented by letter informatically. Therefore in the genetic code we could find the one-letter code.

Secondary structures result from the spatial arrangement of aminoacids that interact with each closer neighbors. There are some remarkable secondary structures such as $\beta$-sheets, $\alpha$-helix, coils (felxible) and others.

Terciary structure informs about the structural disposition of the secondary structures that fold between each other due to hidrophobic interactions, disulphure bonds, and other chemical interactions forming a globular and **dyanamic** structure. Thus, proteins could display multiple structural states depending on the physical and energy stability (see the Levingthan paradox).

Cuaternary structures result from interaction of multiple terciary structures. *The structure, therefore dictates the protein function*. This basic concept have triggered more recently a boom on the analysis of the structure of proteins.

## Identifying or predicting protein structures

Xray crystallography, 
nuclear magnetic resonance (NMR) allows to encapsulated dynamic information of the protein in time
Electron micrography (EM) and cryo-EM. These experiment rely on highly specilized

::: {.callout-note}
To date helium is scarse and many laboratories 
:::

Modelling the structure whether *ab initio* ot by *homology* also allow structure prediction. However these strategies 

Recently AlphaFold

The protein database (PDB) 

Protein topologies resulting from the folding: horshoe, beta-barrel and other could be identified 

Structural classification of proteins (SCOP) when analysing a new protein
classification by class, architecture, topology (fold-family), homologous superfamily and sequence family

Importance of Gene Ontology

### Secunddary structure prediction

Secundary structures could also be represented in one letter (e.g.)

Functional domains could be predicted by sequence alignment and allow structural inference. Main predictions are based heavily on machine learning and are frequently accepted under a consensus of multiple tools.

::: {.callout-warning}
There are a group of protein called intrinsically-disordered proteins that change in its terciary structures quite frequently, therefore it prediction could be troublesome.
:::

::: {.callout-warning}
## Exercise 01
Submit 5lWM (JAK3) from the PDB on FASTA format on the [JPRED](http://www.compbio.dundee.ac.uk/jpred/) and [PSIPRED](http://bioinf.cs.ucl.ac.uk/psipred/) and compared with the [experimentally predicted](https://www.rcsb.org/sequence/5LWM) version of the protein.
Analyse the predicitons and tell are there differences between predictions? Which one is more accurate?
:::

## PDB database introduction

The PDB database is one of the most important and ancient open biological database where all new protein structures are submitted. It is an international consortium where several regions work together to curate infromation.

The PDB in europe [PDBe](pdbe.org) is not only for proteins but form many other *experimentally predicted* macromolecules (protein-protein interactions, peptides, RNA and so on). 

Protein structures are registered using a unique code of four characters.

X-rax crystallography: is a chemical state of the macromolecule where it is imobilized, therefore information correspond to one state of the structure, then cristal protein is submitted to an x-ray beam to generate a difraction pattern. 

NMR spectroscopy: captures dynamic information of the protein, but generally resolves small proteins. The principle?

Electro microscopy adapted to cryopreservation allow proteins visualization.