% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  letterpaper,
  DIV=11,
  numbers=noendperiod]{scrreprt}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{241,243,245}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.40,0.46,0.14}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.28,0.35,0.67}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.07,0.07,0.07}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\usepackage[normalem]{ulem}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% Make \paragraph and \subparagraph free-standing
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
\KOMAoption{captions}{tableheading}
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[many]{tcolorbox}}
\@ifpackageloaded{fontawesome5}{}{\usepackage{fontawesome5}}
\definecolor{quarto-callout-color}{HTML}{909090}
\definecolor{quarto-callout-note-color}{HTML}{0758E5}
\definecolor{quarto-callout-important-color}{HTML}{CC1914}
\definecolor{quarto-callout-warning-color}{HTML}{EB9113}
\definecolor{quarto-callout-tip-color}{HTML}{00A047}
\definecolor{quarto-callout-caution-color}{HTML}{FC5300}
\definecolor{quarto-callout-color-frame}{HTML}{acacac}
\definecolor{quarto-callout-note-color-frame}{HTML}{4582ec}
\definecolor{quarto-callout-important-color-frame}{HTML}{d9534f}
\definecolor{quarto-callout-warning-color-frame}{HTML}{f0ad4e}
\definecolor{quarto-callout-tip-color-frame}{HTML}{02b875}
\definecolor{quarto-callout-caution-color-frame}{HTML}{fd7e14}
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\renewcommand*\contentsname{Table of contents}
\renewcommand*\listfigurename{List of Figures}
\renewcommand*\listtablename{List of Tables}
\renewcommand*\figurename{Figure}
\renewcommand*\tablename{Table}
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother
\makeatletter
\makeatother
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Fundamentals of computational biology},
  pdfauthor={Camilo García-Botero},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}

\title{Fundamentals of computational biology}
\usepackage{etoolbox}
\makeatletter
\providecommand{\subtitle}[1]{% add subtitle to \maketitle
  \apptocmd{\@title}{\par {\large #1 \par}}{}{}
}
\makeatother
\subtitle{Lecture notes}
\author{Camilo García-Botero}
\date{2022-04-11}

\begin{document}
\maketitle

\renewcommand*\contentsname{Table of contents}
{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{preface}{%
\chapter*{Preface}\label{preface}}
\addcontentsline{toc}{chapter}{Preface}

We started this book with the aim of compiling the lectures of the
course Fundamentals of Computational Biology offered at Universidad
EAFIT for undergrad students in Biology. The course has been taught from
different perspectives from its creation, yet the last iteration was
divided into three modules. i) introduction to Unix (4 lectures) ii)
introduction to sequence analysis and genomics (7 lectures) and iii)
principles of structural biology (4 lectures).

Lectures are focused on a theoretical-practical approach were basic
concepts from biology, bioinformatics and computer science and
interleave with the practice to solve challenges.

\hypertarget{introduction}{%
\chapter*{Introduction}\label{introduction}}
\addcontentsline{toc}{chapter}{Introduction}

Here we present a course centered book of the Fundamentals of
Computational Biology. We will cover several topics, from using the unix
tools, the importance of package manager systems (such as homebrew and
conda), sequencing technologies, sequence alignments, molecular
phylogenetics, genome assembly and annotation, and variant calling
analysis.

\part{Unix}

\hypertarget{welcome-to-the-command-line}{%
\chapter{Welcome to the command
line}\label{welcome-to-the-command-line}}

In this chapter we will explore the fundamentals of the command line.
That is the concepts of Unix based systems the command line (CLI) and
how we can use it to access information programmatically.

\part{Sequence analysis}

\hypertarget{introduction-to-sequence-analysis}{%
\chapter{Introduction to sequence
analysis}\label{introduction-to-sequence-analysis}}

In this chapter we will discuss several about several points of view
about bioinformatics and computational biology and how to get started
with the command line being a biologist, we will further consider
several biological concepts that appear central to understand the
manipulation of biological data.

\hypertarget{endless-debate-bioinformatics-vs.-computational-biology}{%
\section*{Endless debate: bioinformatics vs.~computational
biology}\label{endless-debate-bioinformatics-vs.-computational-biology}}
\addcontentsline{toc}{section}{Endless debate: bioinformatics
vs.~computational biology}

\hypertarget{getting-started-with-the-command-line}{%
\section*{Getting started with the command
line}\label{getting-started-with-the-command-line}}
\addcontentsline{toc}{section}{Getting started with the command line}

\hypertarget{the-duality-of-dna}{%
\section*{The duality of DNA}\label{the-duality-of-dna}}
\addcontentsline{toc}{section}{The duality of DNA}

\hypertarget{the-central-dogma-theory-of-molecular-biology-extended}{%
\section*{\texorpdfstring{The central \sout{dogma} theory of molecular
biology
extended}{The central dogma theory of molecular biology extended}}\label{the-central-dogma-theory-of-molecular-biology-extended}}
\addcontentsline{toc}{section}{The central \sout{dogma} theory of
molecular biology extended}

\hypertarget{sequencing-strategies}{%
\section*{Sequencing strategies}\label{sequencing-strategies}}
\addcontentsline{toc}{section}{Sequencing strategies}

\hypertarget{sequencing-over-time}{%
\section*{Sequencing over time}\label{sequencing-over-time}}
\addcontentsline{toc}{section}{Sequencing over time}

\hypertarget{some-insights-from-sequencing-genomes}{%
\section*{Some insights from sequencing
genomes}\label{some-insights-from-sequencing-genomes}}
\addcontentsline{toc}{section}{Some insights from sequencing genomes}

\hypertarget{sanger-analysis}{%
\chapter{Sanger analysis}\label{sanger-analysis}}

This is a section about the first gen sequencing tech

\hypertarget{databases-exploration}{%
\section*{Databases exploration}\label{databases-exploration}}
\addcontentsline{toc}{section}{Databases exploration}

\hypertarget{sanger-sequencing-methods}{%
\section*{Sanger sequencing methods}\label{sanger-sequencing-methods}}
\addcontentsline{toc}{section}{Sanger sequencing methods}

\hypertarget{the-chain-termination-method}{%
\subsection*{The chain termination
method}\label{the-chain-termination-method}}
\addcontentsline{toc}{subsection}{The chain termination method}

\hypertarget{sanger-with-capillary-electrophoresis}{%
\subsection*{Sanger with capillary
electrophoresis}\label{sanger-with-capillary-electrophoresis}}
\addcontentsline{toc}{subsection}{Sanger with capillary electrophoresis}

\hypertarget{strengths-and-limitations-of-sanger-methods}{%
\subsection*{Strengths and limitations of Sanger
methods}\label{strengths-and-limitations-of-sanger-methods}}
\addcontentsline{toc}{subsection}{Strengths and limitations of Sanger
methods}

\hypertarget{files-from-sanger}{%
\section*{Files from Sanger}\label{files-from-sanger}}
\addcontentsline{toc}{section}{Files from Sanger}

\hypertarget{sanger-processing-workflow}{%
\section*{Sanger processing workflow}\label{sanger-processing-workflow}}
\addcontentsline{toc}{section}{Sanger processing workflow}

\hypertarget{the-16s-rrna-and-its-relevance-for-sequencing}{%
\section*{The 16S rRNA and its relevance for
sequencing}\label{the-16s-rrna-and-its-relevance-for-sequencing}}
\addcontentsline{toc}{section}{The 16S rRNA and its relevance for
sequencing}

\part{Challenges demostrations}

\hypertarget{genome-searching}{%
\chapter{Genome searching}\label{genome-searching}}

\begin{tcolorbox}[standard jigsaw,toprule=.15mm, titlerule=0mm, toptitle=1mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, arc=.35mm, bottomrule=.15mm, bottomtitle=1mm, opacityback=0, colback=white, coltitle=black, leftrule=.75mm, rightrule=.15mm, left=2mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, opacitybacktitle=0.6]
Your profesor is interested on knowing how many complete genomes of
Bacillus subtilis are there in the NCBI databases. He ask you later to
count the number of features (genes, CDS, ncRNA, rRNA, etc.) in the
genome of Bacillus subtilis NCIB 3610 (GCF\_002055965.1). And tell you
to document each of the steps and how did you end up with the answer.
Saving the file with your initials (e.g.,
CG-activity01.\{md,txt,docx\})z
\end{tcolorbox}

\hypertarget{downloading-a-genome}{%
\subsection*{Downloading a genome}\label{downloading-a-genome}}
\addcontentsline{toc}{subsection}{Downloading a genome}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{ncbi{-}genome{-}download}
\end{Highlighting}
\end{Shaded}

\hypertarget{downloading-from-ncbi}{%
\subsection*{Downloading from NCBI}\label{downloading-from-ncbi}}
\addcontentsline{toc}{subsection}{Downloading from NCBI}

The first step in this journey is to download a bunch of sequences
programatically. To do so, we will use the program
\href{https://github.com/kblin/ncbi-genome-download}{ncbi-genome-download}.

You could inspect all the options it provides, now we will set our
command as the following:

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{ngd} \AttributeTok{{-}{-}genera} \StringTok{"Bacillus subtilis"}\DataTypeTok{\textbackslash{}}
    \AttributeTok{{-}s}\NormalTok{ refseq}\DataTypeTok{\textbackslash{}}
    \AttributeTok{{-}l}\NormalTok{ complete}\DataTypeTok{\textbackslash{}}
    \AttributeTok{{-}o}\NormalTok{ Data}\DataTypeTok{\textbackslash{}}
    \AttributeTok{{-}{-}flat{-}output}\DataTypeTok{\textbackslash{}}
    \AttributeTok{{-}{-}format}\NormalTok{ features}\DataTypeTok{\textbackslash{}}
    \AttributeTok{{-}n}\NormalTok{ bacteria}\DataTypeTok{\textbackslash{}}
    \KeywordTok{|} \FunctionTok{head} \AttributeTok{{-}n}\NormalTok{ 10}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Considering the following 193 assemblies for download:
GCF_000772125.1 Bacillus subtilis   ATCC 13952
GCF_000772165.1 Bacillus subtilis   ATCC 19217
GCF_000772205.1 Bacillus subtilis   Bs-916
GCF_000782835.1 Bacillus subtilis   SG6
GCF_000789295.1 Bacillus subtilis   PS832
GCF_000952895.1 Bacillus subtilis   BS34A
GCF_000953615.1 Bacillus subtilis   BS49
GCF_001015095.1 Bacillus subtilis   UD1022
GCF_001037985.1 Bacillus subtilis   TO-A JPC
\end{verbatim}

\hypertarget{listing-files}{%
\subsection*{Listing files}\label{listing-files}}
\addcontentsline{toc}{subsection}{Listing files}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ls}\NormalTok{ Data }\KeywordTok{|} \FunctionTok{head} \AttributeTok{{-}n}\NormalTok{ 10}
\end{Highlighting}
\end{Shaded}

\hypertarget{decompressing-using-gzip}{%
\subsection*{\texorpdfstring{Decompressing using
\texttt{gzip}}{Decompressing using gzip}}\label{decompressing-using-gzip}}
\addcontentsline{toc}{subsection}{Decompressing using \texttt{gzip}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{gzip} \AttributeTok{{-}d} \PreprocessorTok{*}
\end{Highlighting}
\end{Shaded}

. . .

\hypertarget{some-files-in-our-data-dir}{%
\subsubsection*{Some files in our data
dir}\label{some-files-in-our-data-dir}}
\addcontentsline{toc}{subsubsection}{Some files in our data dir}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ls}\NormalTok{ Data }\KeywordTok{|} \FunctionTok{head}
\end{Highlighting}
\end{Shaded}

\hypertarget{importing-the-files-into-r}{%
\subsection*{Importing the files into
R}\label{importing-the-files-into-r}}
\addcontentsline{toc}{subsection}{Importing the files into R}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(tidyverse)}
\FunctionTok{library}\NormalTok{(fs)}

\NormalTok{all\_features }\OtherTok{\textless{}{-}} \FunctionTok{dir\_ls}\NormalTok{(}\StringTok{"Data/"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{map\_df}\NormalTok{(read\_tsv)}

\NormalTok{all\_features }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{head}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

. . .

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(tidyverse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
-- Attaching packages --------------------------------------- tidyverse 1.3.1 --
\end{verbatim}

\begin{verbatim}
v ggplot2 3.3.5     v purrr   0.3.4
v tibble  3.1.6     v dplyr   1.0.8
v tidyr   1.2.0     v stringr 1.4.0
v readr   2.1.2     v forcats 0.5.1
\end{verbatim}

\begin{verbatim}
-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(fs)}

\NormalTok{all\_features }\OtherTok{\textless{}{-}} \FunctionTok{dir\_ls}\NormalTok{(}\StringTok{"Data/"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{map\_df}\NormalTok{(read\_tsv)}

\NormalTok{all\_features }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{head}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 0 x 0
\end{verbatim}

\hypertarget{data-processing}{%
\subsection*{Data processing}\label{data-processing}}
\addcontentsline{toc}{subsection}{Data processing}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{all\_features\_grouped }\OtherTok{\textless{}{-}}\NormalTok{ all\_features }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{feature =} \StringTok{\textasciigrave{}}\AttributeTok{\# feature}\StringTok{\textasciigrave{}}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(assembly, feature) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{group\_by}\NormalTok{(assembly, feature) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{  operations}
  \FunctionTok{count}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{pivot\_wider}\NormalTok{(}\AttributeTok{names\_from =}\NormalTok{ feature, }\AttributeTok{values\_from =}\NormalTok{ n) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{arrange}\NormalTok{(}\FunctionTok{desc}\NormalTok{(CDS))}
    
\NormalTok{all\_features\_grouped }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{head}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

create a new dataset that will group by features per accession. get read
of the weird name of the column. Select these two columns. Group by
these two columns to perform. count the numbers of rows based on the
applied group. generate a wide dataset sending row names as columns.
Arrange descending by the number of CDSs.

\hypertarget{sanger-processing}{%
\chapter{Sanger processing}\label{sanger-processing}}

\begin{tcolorbox}[standard jigsaw,toprule=.15mm, titlerule=0mm, toptitle=1mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, arc=.35mm, bottomrule=.15mm, bottomtitle=1mm, opacityback=0, colback=white, coltitle=black, leftrule=.75mm, rightrule=.15mm, left=2mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, opacitybacktitle=0.6]
Your professor gives you a couple of .ab1 files of a 16S rRNA gene from
an old project a student conducted. She tells you to process and analyse
them using the sanger sequence pipeline analyis. And as she doesn't know
from which species they belong, she ask you to identify the organism to
whom it belongs by using the resulting consensus sequence. She finally
reminds you to document each step of the process including the
identification step.
\end{tcolorbox}

\hypertarget{procesing-a-single-.ab1-pair}{%
\section*{\texorpdfstring{Procesing a single \texttt{.ab1}
pair}{Procesing a single .ab1 pair}}\label{procesing-a-single-.ab1-pair}}
\addcontentsline{toc}{section}{Procesing a single \texttt{.ab1} pair}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(sangeranalyseR)}

\NormalTok{groEL }\OtherTok{\textless{}{-}} \FunctionTok{SangerAlignment}\NormalTok{(}
  \AttributeTok{ABIF\_Directory =} \StringTok{"\textasciitilde{}/Projects/Bacillus/Data/Sanger/Inter/groEL/"}\NormalTok{,}
  \AttributeTok{REGEX\_SuffixForward =} \StringTok{"\_1\_F.ab1"}\NormalTok{,}
  \AttributeTok{REGEX\_SuffixReverse =} \StringTok{"\_2\_R.ab1"}\NormalTok{,}
  \AttributeTok{TrimmingMethod =} \StringTok{"M2"}\NormalTok{,}
  \AttributeTok{M2CutoffQualityScore =} \DecValTok{33}\NormalTok{,}
  \AttributeTok{M2SlidingWindowSize =} \DecValTok{10}
\NormalTok{)}

\FunctionTok{writeFasta}\NormalTok{(groEL, }
           \AttributeTok{outputDir =} \StringTok{"\textasciitilde{}/Documents/Teaching/BiologyCourses/BI0487/Demos/02{-}demo{-}sangeranalysis"}\NormalTok{,}
           \AttributeTok{selection =} \StringTok{"contigs\_unalignment"}\NormalTok{,}
\NormalTok{           )}

\FunctionTok{launchApp}\NormalTok{(groEL)}
\FunctionTok{generateReport}\NormalTok{(groEL)}
\FunctionTok{qualityBasePlot}\NormalTok{(groEL)}
\end{Highlighting}
\end{Shaded}

\hypertarget{processing-a-bulk-of-.ab1-files}{%
\section*{\texorpdfstring{Processing a bulk of \texttt{.ab1}
files}{Processing a bulk of .ab1 files}}\label{processing-a-bulk-of-.ab1-files}}
\addcontentsline{toc}{section}{Processing a bulk of \texttt{.ab1} files}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(fs)}
\FunctionTok{library}\NormalTok{(purrr)}

\NormalTok{dirs }\OtherTok{\textless{}{-}}\NormalTok{ fs}\SpecialCharTok{::}\FunctionTok{dir\_ls}\NormalTok{(}\StringTok{"\textasciitilde{}/Projects/Bacillus/Data/Sanger/Inter"}\NormalTok{)}

\NormalTok{sanger\_bulk }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(dir) \{}
  \FunctionTok{SangerAlignment}\NormalTok{(}
    \AttributeTok{ABIF\_Directory =}\NormalTok{ dir,}
    \AttributeTok{REGEX\_SuffixForward =} \StringTok{"\_1\_F.ab1"}\NormalTok{,}
    \AttributeTok{REGEX\_SuffixReverse =} \StringTok{"\_2\_R.ab1"}
\NormalTok{  )}
\NormalTok{\}}

\NormalTok{genes }\OtherTok{\textless{}{-}}\NormalTok{ dirs }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{map}\NormalTok{(sanger\_bulk)}

\FunctionTok{launchApp}\NormalTok{(genes}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{/Users/camilogarcia/Projects/Bacillus/Data/Sanger/Inter/gyrA}\StringTok{\textasciigrave{}}\NormalTok{)}

\FunctionTok{writeFasta}\NormalTok{(}
  \AttributeTok{outputDir =} \StringTok{"\textasciitilde{}/Documents/Teaching/BiologyCourses/BI0487/Demos/02{-}demo{-}sangeranalysis"}\NormalTok{,}
  \AttributeTok{selection =} \StringTok{"contigs\_unalignment"}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{sequence-alignment-demo}{%
\chapter{Sequence alignment demo}\label{sequence-alignment-demo}}

\begin{tcolorbox}[standard jigsaw,toprule=.15mm, titlerule=0mm, toptitle=1mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, arc=.35mm, bottomrule=.15mm, bottomtitle=1mm, opacityback=0, colback=white, coltitle=black, leftrule=.75mm, rightrule=.15mm, left=2mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, opacitybacktitle=0.6]
Your professor is working with species from genus Bacillus and want to
align an orthologous gene from 10 genomes of different isolates. He
gives you the GenBank accession number of these isolates and ask you to
select one orthologus gene (Nucleotide seq) that you consider might be
useful to differentiate the bacterial isolates and ask you to align
those genes as you better consider. He finally ask you to document each
step and send him the sequence alignment file in FASTA format along with
the sequence alignment general stats in a TXT file (length, number of
each nucletides and other stats you consider important).

Accessions: GCA\_012225885.1, GCA\_000196735.1, GCA\_000742895.1,
GCA\_001584335.1, GCA\_000007825.1, GCA\_000832905.1, GCA\_000008425.1,
GCA\_000507105.1, GCA\_000832605.1, GCA\_900186955.1
\end{tcolorbox}

\hypertarget{download-sequences}{%
\section*{Download sequences}\label{download-sequences}}
\addcontentsline{toc}{section}{Download sequences}

Make sure to use the \texttt{-\/-flat-output} avoiding download of
multiple metadata

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{ngd} \AttributeTok{{-}{-}flat{-}output} \AttributeTok{{-}p}\NormalTok{ 4 }\AttributeTok{{-}s}\NormalTok{ genbank }\AttributeTok{{-}A}\NormalTok{ genome{-}accessions.txt }\AttributeTok{{-}F}\NormalTok{ cds{-}fasta bacteria}
\end{Highlighting}
\end{Shaded}

In this case \texttt{cds-fasta} parameter will download the nucleotide
sequences of the gene. Other alternatives could be useful such as blast
search on a genome database or searching through the GENBANK annotation
files (both files also could be downloaded using \texttt{ngd}).

\hypertarget{unwrapping-fasta-records}{%
\section*{Unwrapping FASTA records}\label{unwrapping-fasta-records}}
\addcontentsline{toc}{section}{Unwrapping FASTA records}

NCBI registries came with an undesirable wrapping around the lines of
sequencing which basically is inserting a return character after some
established number of characters. Then a way to get rid of them is to
use a command line utility from
\href{https://github.com/AstrobioMike}{AstrobioMike (Mike Lee)} which
will give a line per sequence after the FASTA header. We can later
assume the the first line after the header will be the entire sequence

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{for}\NormalTok{ i }\KeywordTok{in}\NormalTok{ GCA\_}\PreprocessorTok{*}\KeywordTok{;} \ControlFlowTok{do} 
    \VariableTok{N}\OperatorTok{=}\VariableTok{$(}\FunctionTok{basename} \VariableTok{$i}\NormalTok{ .fna}\VariableTok{)}\KeywordTok{;} 
    \FunctionTok{bash}\NormalTok{ bit{-}remove{-}wraps.sh }\VariableTok{$\{i\}} \OperatorTok{\textgreater{}} \VariableTok{$\{N\}}\NormalTok{\_unwrapped.fasta}\KeywordTok{;} 
\ControlFlowTok{done}
\end{Highlighting}
\end{Shaded}

\hypertarget{gene-search}{%
\section*{Gene search}\label{gene-search}}
\addcontentsline{toc}{section}{Gene search}

A possible way to search throughout the file registries is by using the
\texttt{grep} command, that recursively will search each file. Fine
tuned it allow to search for the first match, but also for the
``after-context'' in terms of lines desired to be printed:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep} \AttributeTok{{-}h}\DataTypeTok{\textbackslash{}}
     \AttributeTok{{-}m}\NormalTok{ 1}\DataTypeTok{\textbackslash{}}
     \AttributeTok{{-}A}\NormalTok{ 1}\DataTypeTok{\textbackslash{}}
     \AttributeTok{{-}E} \StringTok{"DNA gyrase, A| gyrase subunit A | gyrase alpha| gyrase \textbackslash{}(subunit A\textbackslash{})| gyrA"}  \PreprocessorTok{*}\NormalTok{.fasta }\KeywordTok{|}\DataTypeTok{\textbackslash{}}
     \FunctionTok{sed} \StringTok{"s/{-}{-}//g"} \KeywordTok{|}\DataTypeTok{\textbackslash{}}
     \FunctionTok{sed} \StringTok{"/\^{}$/d"}
\end{Highlighting}
\end{Shaded}

After finding the genes we could exclude some lines using \texttt{sed}
avoid the ``--'' characters and the empty blank line using the
appropriate regular expression (\^{}\$/d) . We are now with an almost
clean multi sequence file, because header names are still and will be
problematic. How do we programatically change the FASTA headers? We will
see in the next step.

\hypertarget{renaming-fasta-headers}{%
\section*{Renaming fasta headers}\label{renaming-fasta-headers}}
\addcontentsline{toc}{section}{Renaming fasta headers}

A simple but powerful script to do this is
\texttt{bit-dedup-fasta-heades} it was developed by
\href{https://github.com/AstrobioMike}{AstrobioMike (Mike Lee)} and it
simply parses the headers and substitutes by a simple encoder found en
each of them:

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{python}\NormalTok{ bit{-}dedupe{-}fasta{-}headers.py }\AttributeTok{{-}i}\NormalTok{ all\_gyrA.fasta }\AttributeTok{{-}o}\NormalTok{ all\_gyrA\_renamed.fasta}
\end{Highlighting}
\end{Shaded}

Now the the files has files names that are simply to work with. Which
will enable to asses better out sequence alignment matrix.

\hypertarget{sequence-alignment}{%
\section*{Sequence alignment}\label{sequence-alignment}}
\addcontentsline{toc}{section}{Sequence alignment}

There are many programs that are suited for performed multiple sequence
alignments. Perhaps the two most used are
\href{https://mafft.cbrc.jp/alignment/software/}{\texttt{MAFFT}} and
\href{https://drive5.com/muscle5/}{\texttt{MUSCLE}} both specialized in
multiple sequence alignment (that is: when having two or more than two
sequences). The second tends to be more accurate when having large
data-sets, but the first on is more versatile, fast and accurate on
different kind of data-sets.

~~~~Both programas take as input a single file containing all the
sequences concatenated horizontally (that is a multi-fasta file)
careless of the extension but (MFA, FA, FASTA, FNA, etc). And generate a
simple output (whether with the \texttt{-o} in \texttt{MUSCLE} or to the
std output in \texttt{MAFFT}) ~~~~

\begin{Shaded}
\begin{Highlighting}[]

\ExtensionTok{ginsi} \AttributeTok{{-}{-}preservecase} \AttributeTok{{-}{-}reorder}\NormalTok{ all\_gyrA\_renamed.fasta }\OperatorTok{\textgreater{}}\NormalTok{ all\_gyrA\_renamed\_ginsi.fasta }\CommentTok{\# globally optimized}
\ExtensionTok{einsi} \AttributeTok{{-}{-}preservecase} \AttributeTok{{-}{-}reorder}\NormalTok{ all\_gyrA\_renamed.fasta }\OperatorTok{\textgreater{}}\NormalTok{ all\_gyrA\_renamed\_einsi.fasta }\CommentTok{\# gene{-}focused optimized}
\ExtensionTok{linsi} \AttributeTok{{-}{-}preservecase} \AttributeTok{{-}{-}reorder}\NormalTok{ all\_gyrA\_renamed.fasta }\OperatorTok{\textgreater{}}\NormalTok{ all\_gyrA\_renamed\_linsi.fasta }\CommentTok{\# locally optimized}

\ExtensionTok{muscle} \AttributeTok{{-}i}\NormalTok{ all\_gyrA\_renamed.fasta }\AttributeTok{{-}o}\NormalTok{ all\_gyrA\_renamed\_muscle.fasta}

\ExtensionTok{famsa} \AttributeTok{{-}t}\NormalTok{ 8 all\_gyrA\_renamed.fasta }\OperatorTok{\textgreater{}}\NormalTok{ all\_gyrA\_renamed\_famsa.fasta}

\ExtensionTok{kalign} \AttributeTok{{-}i}\NormalTok{ all\_gyrA\_renamed.fasta }\AttributeTok{{-}o}\NormalTok{ all\_gyrA\_renamed\_kalign.fasta}
\end{Highlighting}
\end{Shaded}

\hypertarget{assesment-of-the-alignment}{%
\section*{Assesment of the alignment}\label{assesment-of-the-alignment}}
\addcontentsline{toc}{section}{Assesment of the alignment}

Inspection of the alignment is there very first step for assesing its
quality. A CDS tends to genereate a codon-like alignment starting with
the methione codon (ATG,GTG) and finishing with a stop (TAA, TAG, etc.).
Therefore finding this structure when aligning a complete genes is
expected. If a middle fraction of the gene is being aligned ORF might
not display any stop codon. Verifying a codon-like alignment shows a
biological order on the sequences other that mere artifact of the
alignment, that is an evolutionary behavior of the sequence. We can do
it usin \href{https://telatin.github.io/seqfu2/}{seqfu} from the CLI or
interactively with \href{https://github.com/AliView/AliView}{AliView}.

A second step is to find the variability of the alignment. A simple way
to find that is to calculate simpl stats from the alignment (sites,
variable sites, As, Ts, etc.). A powerful cli program to do so is
\href{https://github.com/evolbioinfo/goalign}{goalign}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{goalign}\NormalTok{ stats }\AttributeTok{{-}i}\NormalTok{ all\_gyrA\_renamed\_linsi.fasta}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{length}\NormalTok{  2508}
\ExtensionTok{nseqs}\NormalTok{   8}
\ExtensionTok{avgalleles}\NormalTok{  1.7400}
\ExtensionTok{variable}\NormalTok{ sites  1202}
\ExtensionTok{char}\NormalTok{    nb  freq}
\ExtensionTok{{-}}\NormalTok{   273 0.013606}
\ExtensionTok{A}\NormalTok{   6418    0.319876}
\ExtensionTok{C}\NormalTok{   3633    0.181071}
\ExtensionTok{G}\NormalTok{   4755    0.236992}
\ExtensionTok{T}\NormalTok{   4985    0.248455}
\ExtensionTok{alphabet}\NormalTok{    nucleotide}
\end{Highlighting}
\end{Shaded}

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

\hypertarget{an-alternative-approach-using-blast}{%
\section*{An alternative approach using
BLAST}\label{an-alternative-approach-using-blast}}
\addcontentsline{toc}{section}{An alternative approach using BLAST}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{ngd} \AttributeTok{{-}{-}flat{-}output} \AttributeTok{{-}p}\NormalTok{ 4 }\AttributeTok{{-}s}\NormalTok{ genbank }\AttributeTok{{-}A}\NormalTok{ genome{-}accessions.txt }\AttributeTok{{-}F}\NormalTok{ fasta }\AttributeTok{{-}{-}parallel}\NormalTok{ 8 bacteria}

\ControlFlowTok{for}\NormalTok{ i }\KeywordTok{in}\NormalTok{ GCA\_}\PreprocessorTok{*}\KeywordTok{;} \ControlFlowTok{do} \FunctionTok{cat} \VariableTok{$\{i\}} \OperatorTok{\textgreater{}\textgreater{}}\NormalTok{ all\_genomes.fasta}\KeywordTok{;} \ControlFlowTok{done}

\ExtensionTok{makeblastdb} \AttributeTok{{-}in}\NormalTok{ all\_genomes.fasta }\AttributeTok{{-}parse\_seqids} \AttributeTok{{-}blastdb\_version}\NormalTok{ 5 }\AttributeTok{{-}title} \StringTok{"demo"} \AttributeTok{{-}dbtype}\NormalTok{ nucl}

\ExtensionTok{blastn} \AttributeTok{{-}db}\NormalTok{ all\_genomes.fasta }\AttributeTok{{-}query}\NormalTok{ gyrA.fasta }\AttributeTok{{-}outfmt} \StringTok{"6 sseqid sseq"} \AttributeTok{{-}word\_size}\NormalTok{ 5 }\AttributeTok{{-}evalue}\NormalTok{ 1e{-}50 }\AttributeTok{{-}num\_threads}\NormalTok{ 8 }\KeywordTok{|} \FunctionTok{sed} \AttributeTok{{-}E} \StringTok{"s/gb|emb/\textgreater{}/g"} \KeywordTok{|} \FunctionTok{sed} \AttributeTok{{-}E} \StringTok{"s/\textbackslash{}t/\textbackslash{}n/"} \OperatorTok{\textgreater{}}\NormalTok{ all\_gyrA.fasta}
\end{Highlighting}
\end{Shaded}

\hypertarget{the-alternative-using-the-gcf}{%
\section*{The alternative using the
GCF}\label{the-alternative-using-the-gcf}}
\addcontentsline{toc}{section}{The alternative using the GCF}

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{for}\NormalTok{ i }\KeywordTok{in} \PreprocessorTok{*}\NormalTok{fna}\KeywordTok{;} \ControlFlowTok{do}\KeywordTok{;} \ExtensionTok{goalign}\NormalTok{ subset }\AttributeTok{{-}e} \StringTok{"gyrA"} \AttributeTok{{-}i} \VariableTok{$\{i\}} \AttributeTok{{-}{-}unaligned}\KeywordTok{;}\ControlFlowTok{done} \KeywordTok{|} \FunctionTok{grep} \StringTok{"\textgreater{}"}
\end{Highlighting}
\end{Shaded}

\hypertarget{phylogenetics-reconstruction}{%
\chapter{Phylogenetics
reconstruction}\label{phylogenetics-reconstruction}}

\begin{tcolorbox}[standard jigsaw,toprule=.15mm, titlerule=0mm, toptitle=1mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, arc=.35mm, bottomrule=.15mm, bottomtitle=1mm, opacityback=0, colback=white, coltitle=black, leftrule=.75mm, rightrule=.15mm, left=2mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, opacitybacktitle=0.6]
Your professor has been working with some mammal species and want to
know the relationships of some sampled individuals. To do so he
extracted the DNA and amplified the mitochondrial CYTB gene of those
individuals. He gives a folder with multiple sequences and ask you to
align them and to reconstruct two trees one using maximum likelihood
(ML) and other using a Bayesian inference (BI). Then ask you to explain
if both trees are congruent with each other.
\end{tcolorbox}

\hypertarget{sequence-alignment-cytb}{%
\section*{\texorpdfstring{Sequence alignment
\emph{cytb}}{Sequence alignment cytb}}\label{sequence-alignment-cytb}}
\addcontentsline{toc}{section}{Sequence alignment \emph{cytb}}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{linsi} \AttributeTok{{-}{-}preservecase} \AttributeTok{{-}{-}reorder}\NormalTok{ cytB.fasta }\OperatorTok{\textgreater{}}\NormalTok{ cytB{-}aligned.fasta}
\end{Highlighting}
\end{Shaded}

\hypertarget{evolutionary-substitution-model}{%
\section*{Evolutionary substitution
model}\label{evolutionary-substitution-model}}
\addcontentsline{toc}{section}{Evolutionary substitution model}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{modeltest{-}ng} \AttributeTok{{-}i}\NormalTok{ cytB{-}aligned.fasta }\AttributeTok{{-}d}\NormalTok{ nt }\AttributeTok{{-}o}\NormalTok{ model{-}cytB.txt}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{04{-}demo{-}phylogenetics/model{-}cytB.txt.log}
\end{Highlighting}
\end{Shaded}

\hypertarget{maximum-likelihood-reconstruction}{%
\section*{Maximum likelihood
reconstruction}\label{maximum-likelihood-reconstruction}}
\addcontentsline{toc}{section}{Maximum likelihood reconstruction}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{raxml{-}ng} \AttributeTok{{-}{-}msa}\NormalTok{ cytB{-}aligned.fasta }\AttributeTok{{-}{-}model}\NormalTok{ GTR+I+G4}
\end{Highlighting}
\end{Shaded}

\hypertarget{tree-building}{%
\subsection*{Tree building}\label{tree-building}}
\addcontentsline{toc}{subsection}{Tree building}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{raxml\_data }\OtherTok{\textless{}{-}} \FunctionTok{read.tree}\NormalTok{(}\StringTok{"cytB{-}aligned.fasta.raxml.support"}\NormalTok{)}

\NormalTok{raxml\_data}\SpecialCharTok{$}\NormalTok{tip.label }\OtherTok{\textless{}{-}} \FunctionTok{str\_replace\_all}\NormalTok{(raxml\_data}\SpecialCharTok{$}\NormalTok{tip.label, }\StringTok{"\_"}\NormalTok{, }\StringTok{" "}\NormalTok{)}

\NormalTok{(}
\NormalTok{  raxml\_tree }\OtherTok{\textless{}{-}} \FunctionTok{ggtree}\NormalTok{(raxml\_data) }\SpecialCharTok{+}
    \FunctionTok{geom\_tiplab}\NormalTok{() }\SpecialCharTok{+}
    \CommentTok{\# geom\_point2(aes(subset = !isTip, fill = as.integer(label)), shape = 24, size = 3) +}
    \FunctionTok{geom\_text}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{label =} \FunctionTok{as.integer}\NormalTok{(label))) }\SpecialCharTok{+}
    \FunctionTok{theme\_tree}\NormalTok{(}\AttributeTok{legend.position =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.8}\NormalTok{, }\FloatTok{0.7}\NormalTok{)) }
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{bayes-inference-reconstruction}{%
\section*{Bayes inference
reconstruction}\label{bayes-inference-reconstruction}}
\addcontentsline{toc}{section}{Bayes inference reconstruction}

\hypertarget{tree-building-1}{%
\subsection*{Tree building}\label{tree-building-1}}
\addcontentsline{toc}{subsection}{Tree building}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tail} \AttributeTok{{-}n}\NormalTok{ 18 ../Data/cytB{-}mb.nex}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{mb} \AttributeTok{{-}i}\NormalTok{ ../../Data}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{sumtrees.py} \AttributeTok{{-}s}\NormalTok{ mcct }\AttributeTok{{-}o}\OperatorTok{=}\NormalTok{cytB{-}mb{-}mcct.tre cytB{-}aligned.fasta{-}out.nex.run1.t}
\end{Highlighting}
\end{Shaded}

\hypertarget{sequence-reads-assesment}{%
\chapter{Sequence reads assesment}\label{sequence-reads-assesment}}

\begin{tcolorbox}[standard jigsaw,toprule=.15mm, titlerule=0mm, toptitle=1mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, arc=.35mm, bottomrule=.15mm, bottomtitle=1mm, opacityback=0, colback=white, coltitle=black, leftrule=.75mm, rightrule=.15mm, left=2mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, opacitybacktitle=0.6]
Your professor challenges you to assembly a bacterial genome. He wants
to know if an assembly using Illumina reads or Nanopore reads is better.
The raw sequences for Illumina could be downloaded with the code
SRR15634574. And the Nanopore raw reads with SRR15634573 here:
https://sra-explorer.info. Choose at least on set and follow the
instruction from the lecture to assemble the genome. Document each step
and send to the professor the \texttt{assembly-\{your-initials\}.fasta}.
You could form groups of max. 4 students.
\end{tcolorbox}

\hypertarget{general-stats-from-fastq-files}{%
\section*{General stats from fastq
files}\label{general-stats-from-fastq-files}}
\addcontentsline{toc}{section}{General stats from fastq files}

A simple but fast cli to display the general stats from fastq is
\texttt{seqfu}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{seqfu}\NormalTok{ stats –n }\PreprocessorTok{*}\NormalTok{.gz}
\end{Highlighting}
\end{Shaded}

\hypertarget{a-graphical-assessment-of-reads}{%
\section*{A graphical assessment of
reads}\label{a-graphical-assessment-of-reads}}
\addcontentsline{toc}{section}{A graphical assessment of reads}

Often its important to assess reads graphically and \texttt{nanoplot}
offers a complete graphical summary (and general stats as well) of the
reads

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{nanoplot}\NormalTok{ –{-}fastq reads.fastq.gz –{-}output reads{-}report}
\end{Highlighting}
\end{Shaded}

\hypertarget{genome-assembly}{%
\chapter{Genome assembly}\label{genome-assembly}}

\begin{tcolorbox}[standard jigsaw,toprule=.15mm, titlerule=0mm, toptitle=1mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, arc=.35mm, bottomrule=.15mm, bottomtitle=1mm, opacityback=0, colback=white, coltitle=black, leftrule=.75mm, rightrule=.15mm, left=2mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, opacitybacktitle=0.6]
Your professor has sequenced a bacterial isolate using PacBio and
Nanopore sequencing methods and has got the FASTQ files from both
technologies. Now he needs to know the quality and quantity of these
data before start any other analysis and ask you to assess the data. He
needs to know how many sequences there are, how many base pairs (in GB)
are there and the N50. He is also interested in see a visualization of
the i) number of bases vs.~sequence lengths (log transformed) and ii)
the read length vs.~read quality vs.~read number.

He ask you to document every step and to conclude what data should be
used.
\end{tcolorbox}

\hypertarget{download-the-reads}{%
\section*{Download the reads}\label{download-the-reads}}
\addcontentsline{toc}{section}{Download the reads}

\hypertarget{downloading-seqs}{%
\label{downloading-seqs}}%
\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{wget}
\end{Highlighting}
\end{Shaded}

\hypertarget{assess-read-qualities}{%
\section*{Assess read qualities}\label{assess-read-qualities}}
\addcontentsline{toc}{section}{Assess read qualities}

When using illumina \texttt{fastqc} is a very fast alternative. For
nanopore \texttt{nanplot} will do the job.

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{fastqc}
\end{Highlighting}
\end{Shaded}

\hypertarget{exploring-assemblers}{%
\section*{Exploring assemblers}\label{exploring-assemblers}}
\addcontentsline{toc}{section}{Exploring assemblers}

One of the most popular genome assemblers for NGS is \texttt{spades}
whereas for TGS \texttt{flye} has been widely used

\hypertarget{shovill-spades-under-the-hood}{%
\subsection*{Shovill: spades under the
hood}\label{shovill-spades-under-the-hood}}
\addcontentsline{toc}{subsection}{Shovill: spades under the hood}

\href{https://github.com/tseemann/shovill}{\texttt{shovill}} is a
pipeline that enables pre and post processing of genomic data. It can be
tunned to several tools for the processing steps and also to select
different standalone assemblers

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{shovill} \AttributeTok{{-}{-}outdir}\NormalTok{ MxanthusIllumina}\DataTypeTok{\textbackslash{}}
        \AttributeTok{{-}{-}R1}\NormalTok{ R1.fq.gz}\DataTypeTok{\textbackslash{}}
        \AttributeTok{{-}{-}R2}\NormalTok{ R2.fq.gz}\DataTypeTok{\textbackslash{}}
        \AttributeTok{{-}{-}trim}\DataTypeTok{\textbackslash{}}
        \AttributeTok{{-}{-}cpus}\NormalTok{ 32}
\end{Highlighting}
\end{Shaded}

\hypertarget{dragonflye-flye-under-the-hood}{%
\subsection*{Dragonflye: flye under the
hood}\label{dragonflye-flye-under-the-hood}}
\addcontentsline{toc}{subsection}{Dragonflye: flye under the hood}

Similar to \texttt{shovill} (and inspired by it)
\href{}{\texttt{dragonflye}} is a pipeline that enables several
processing steps of genomic data be

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{dragonflye} \AttributeTok{{-}{-}outdir}\NormalTok{ MxanthusNanopore}\DataTypeTok{\textbackslash{}}
           \AttributeTok{{-}{-}gsize}\NormalTok{ 9Mb}\DataTypeTok{\textbackslash{}}
           \AttributeTok{{-}{-}trim}\DataTypeTok{\textbackslash{}}
           \AttributeTok{{-}{-}reads}\NormalTok{ ont{-}readsfastq.gz}\DataTypeTok{\textbackslash{}}
           \AttributeTok{{-}{-}racon}\NormalTok{ 5}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{wget}
\end{Highlighting}
\end{Shaded}

Since we are trying to assemble a bacterial genome, computer memory
appears to be a limiting features of a local machine. Then, a computer
cluster with high performance turns out to be an important need.

First we need environment installations, therefore its important to have
conda environments with the assemblers and other programs
(\texttt{conda\ create\ -c\ bioconda\ dragonflye\ dragonflye} and
\texttt{conda\ create\ -c\ bioconda\ shovill\ shovill}). That way both
assemblers pipelines will lie in separate environment avoiding possible
dependenies problems

We will use \href{}{Apolo} computer cluster which uses \href{}{Slurm} as
the computer system workload manager (i.e a program that manages the
time and resources of the computer).

\hypertarget{references}{%
\chapter*{References}\label{references}}
\addcontentsline{toc}{chapter}{References}

\hypertarget{refs}{}
\begin{CSLReferences}{0}{0}
\end{CSLReferences}

\end{document}
