% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  letterpaper,
  DIV=11,
  numbers=noendperiod,
  oneside]{scrreprt}

\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[left=1in,marginparwidth=2.0666666666667in,textwidth=4.1333333333333in,marginparsep=0.3in]{geometry}
\usepackage[normalem]{ulem}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{5}
% Make \paragraph and \subparagraph free-standing
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{241,243,245}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.40,0.45,0.13}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.28,0.35,0.67}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.07,0.07,0.07}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

\KOMAoption{captions}{tableheading}
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[many]{tcolorbox}}
\@ifpackageloaded{fontawesome5}{}{\usepackage{fontawesome5}}
\definecolor{quarto-callout-color}{HTML}{909090}
\definecolor{quarto-callout-note-color}{HTML}{0758E5}
\definecolor{quarto-callout-important-color}{HTML}{CC1914}
\definecolor{quarto-callout-warning-color}{HTML}{EB9113}
\definecolor{quarto-callout-tip-color}{HTML}{00A047}
\definecolor{quarto-callout-caution-color}{HTML}{FC5300}
\definecolor{quarto-callout-color-frame}{HTML}{acacac}
\definecolor{quarto-callout-note-color-frame}{HTML}{4582ec}
\definecolor{quarto-callout-important-color-frame}{HTML}{d9534f}
\definecolor{quarto-callout-warning-color-frame}{HTML}{f0ad4e}
\definecolor{quarto-callout-tip-color-frame}{HTML}{02b875}
\definecolor{quarto-callout-caution-color-frame}{HTML}{fd7e14}
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{bookmark}{}{\usepackage{bookmark}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[many]{tcolorbox}}
\makeatother
\makeatletter
\@ifundefined{shadecolor}{\definecolor{shadecolor}{rgb}{.97, .97, .97}}
\makeatother
\makeatletter
\@ifpackageloaded{sidenotes}{}{\usepackage{sidenotes}}
\@ifpackageloaded{marginnote}{}{\usepackage{marginnote}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{fontawesome5}{}{\usepackage{fontawesome5}}
\makeatother
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Fundamentals of computational biology},
  pdfauthor={Camilo García-Botero},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}

\title{Fundamentals of computational biology}
\usepackage{etoolbox}
\makeatletter
\providecommand{\subtitle}[1]{% add subtitle to \maketitle
  \apptocmd{\@title}{\par {\large #1 \par}}{}{}
}
\makeatother
\subtitle{Lecture notes}
\author{Camilo García-Botero}
\date{2022-08-19}

\begin{document}
\maketitle
\ifdefined\Shaded\renewenvironment{Shaded}{\begin{tcolorbox}[boxrule=0pt, enhanced, borderline west={3pt}{0pt}{shadecolor}, sharp corners, breakable, interior hidden, frame hidden]}{\end{tcolorbox}}\fi

\renewcommand*\contentsname{Table of contents}
{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{2}
\tableofcontents
}
\bookmarksetup{startatroot}

\hypertarget{preface}{%
\chapter*{Preface}\label{preface}}
\addcontentsline{toc}{chapter}{Preface}

We started this book with the aim of compiling the lectures of the
course Fundamentals of Computational Biology offered at Universidad
EAFIT for undergrad students in Biology. The course has been taught from
different perspectives from its creation, yet the last iteration was
divided into five introductory modules to i) Unix, ii) sequence
analysis, iii) genomics, iii) metagenomics iv) structural biology.

Lectures are focused on a theoretical-practical approach where basic
concepts from biology, bioinformatics and computer science and
interleave with the practice to solve challenges. Exercised or
challenges are designed to improve students' abilities that are likely
to be involved in real-life problems in computational biology.

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-caution-color}{\faFire}\hspace{0.5em}{Danger}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-caution-color-frame, colbacktitle=quarto-callout-caution-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
This book is a work in progress. If you find issues, typos or relevant
information to share with, anything is welcome.
\end{tcolorbox}

\hypertarget{learning-features}{%
\section*{Learning features}\label{learning-features}}
\addcontentsline{toc}{section}{Learning features}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-note-color}{\faInfo}\hspace{0.5em}{Note}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-note-color-frame, colbacktitle=quarto-callout-note-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
Sometimes other fields might add interested value to the understanding
of the computational biology area. This feature remarks some of them and
aim to explain these intersections.
\end{tcolorbox}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
As you move forward in the computational biology field you will find
that there are several tips and tricks (mainly from the command line) as
well as some random CLI programs that can leverage your daily workflow
as a researcher. Using this feature we highlight some of those that
appeared to linger on the field.
\end{tcolorbox}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-important-color}{\faExclamation}\hspace{0.5em}{Important}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-important-color-frame, colbacktitle=quarto-callout-important-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
To help you consolidate your understanding we end most chapters with
important messages or concepts that help you evaluate yourself as you
move forward on the lessons.
\end{tcolorbox}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-caution-color}{\faFire}\hspace{0.5em}{Danger}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-caution-color-frame, colbacktitle=quarto-callout-caution-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]

\bookmarksetup{startatroot}

\hypertarget{caution}{%
\chapter*{Caution}\label{caution}}
\addcontentsline{toc}{chapter}{Caution}

When experimenting with the CLI and many other computational tools it is
common to face several known errors and drawbacks. Then, we present some
of them and how to sort them out.

\end{tcolorbox}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Warning}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]

\bookmarksetup{startatroot}

\hypertarget{challenges}{%
\chapter*{Challenges}\label{challenges}}
\addcontentsline{toc}{chapter}{Challenges}

Since focused on a competences learning approach we have highlighted
several real-life (but basic) \emph{challenges} a researcher faces when
approaching computational biology problem (from tool selection, usage
and result analysis). Therefore the book section \emph{challenges}
presents a selection of these problems that will later be approached by
a computational biology strategy (mainly from the CLI).

\end{tcolorbox}

\begin{tcolorbox}[enhanced jigsaw, colback=white, opacityback=0, leftrule=.75mm, breakable, colframe=quarto-callout-color-frame, rightrule=.15mm, arc=.35mm, bottomrule=.15mm, toprule=.15mm, left=2mm]

\bookmarksetup{startatroot}

\hypertarget{file-format}{%
\chapter*{File format}\label{file-format}}
\addcontentsline{toc}{chapter}{File format}

As many analysis specialize on data analysis, many formats arise that
optimize the processing steps or the data storing steps. Some of these
formats are keystones of bioinformatic analyses. We present examples of
some formats an describe its main elements.

\end{tcolorbox}

\bookmarksetup{startatroot}

\hypertarget{introduction}{%
\chapter*{Introduction}\label{introduction}}
\addcontentsline{toc}{chapter}{Introduction}

The present book gathers the lecture notes of the undergrad course in
Fundamentals of Computational Biology that takes place in EAFIT
University. The first part will focus on how to use the the command line
interface (aka CLI). It includes a long-format chapter about the Unix
tools and concepts that is taught during the first four class lessons.
The second chapter of covers the basics of git and the principal
workflows to work daily on a collaborative manner this is actually the
second lesson of the course. the importance of package manager systems
(such as homebrew and conda

The second part We will cover several topics sequencing technologies,
sequence alignments, molecular phylogenetics, genome assembly and
annotation, and variant calling analysis.

Include a section, maybe an appendix about how to handle errors Include
s section about the windows subsystem for linux (WSL) and the ease of
use for windows users

\hypertarget{endless-debate-bioinformatics-vs.-computational-biology}{%
\section*{Endless debate: bioinformatics vs.~computational
biology}\label{endless-debate-bioinformatics-vs.-computational-biology}}
\addcontentsline{toc}{section}{Endless debate: bioinformatics
vs.~computational biology}

\hypertarget{the-extent-of-computational-biology}{%
\section*{The extent of computational
biology}\label{the-extent-of-computational-biology}}
\addcontentsline{toc}{section}{The extent of computational biology}

\part{The command line}

\hypertarget{sec-cli}{%
\chapter{The command line}\label{sec-cli}}

\begin{quote}
In this chapter we will explore the fundamentals of the command line
interface (aka CLI). We will distinguish the differences between Unix,
CLI, Bash and Terminal and other concepts from the computer sciences.

As you will see the CLI is composed of several programs enabling the
interaction with the machine, we will discuss some of the basics to
navigate your machine, and some advance one that enable complex
operations and automating tasks.
\end{quote}

\hypertarget{getting-started-with-the-command-line}{%
\section{Getting started with the command
line}\label{getting-started-with-the-command-line}}

Before landing into the CLI let us consider the Unix concept. The first
question that comes in this section is: what is Unix? It simply is an
\href{https://en.wikipedia.org/wiki/Operating_system}{:operating system
(OS)}. In other words, it is a set of programs that inter-operate with
each other to let you communicate with the machine. A very important
variant (or clone) of Unix is the very well known OS
\href{https://en.wikipedia.org/wiki/Linux}{:Linux}, which was created by
\href{https://en.wikipedia.org/wiki/Linus_Torvalds}{:Linus Torvalds}
from scratch. The most important idea behind Unix based systems is the
idea that we can use it to access information and hardware
programmatically. Other main feature from Unix-like OS systems is the
fact that data is usually stored as text files and the interface by
which users communicate with the machine is also text-based (TUI: text
user interface as opposed to GUI, graphic user interface).

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/terminal.png}

}

\caption{\label{fig-terminal}A \textbf{terminal} app displaying common
features of the command line interface}

\end{figure}

Almost every computer has a way to interact with or access to the inner
elements of the computer. Such interface is called the the
command-line-interface Fig.~\ref{fig-terminal}.

\hypertarget{file-paths}{%
\subsection{File paths}\label{file-paths}}

Programs, files and directories on every machine (with Unix-like OS)
display hierarchical paths (routes), starting out from the \textbf{root}
(represented by the back-slash character \texttt{/}). The \textbf{root}
represents the beginning of all the software installed in the machine.
And many other files are nested from there forming a tree-like structure
for the paths Fig.~\ref{fig-tree}

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/dir-tree.png}

}

\caption{\label{fig-tree}A terminal displaying tree-like structure of
the programs in a machine with macOS}

\end{figure}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]

You can inspect the paths of a nested directory tree using \texttt{tree}
command in you cli:

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{tree} \AttributeTok{{-}d} \AttributeTok{{-}L}\NormalTok{ 1}
\end{Highlighting}
\end{Shaded}

\end{tcolorbox}

There are basically \textbf{two} ways to explore or navigate your file
system. If you always represent it from the root, then you are
presenting an \textbf{absolute path}. For instance the absolute path to
my desktop is (\texttt{/Users/camilogarcia/Desktop}).

\hypertarget{basic-unix-commands}{%
\subsection{Basic Unix commands}\label{basic-unix-commands}}

Given that the vast majority of file systems are organized in file
paths, the first question when starting with the CLI is ``Where am I?''.
So Unix tool system is equipped with a bunch of commands but its basic
ones are pretty much oriented to answer that question and navigating
this text-based interface of files. The following three commands
(\texttt{pwd}, \texttt{cd}, \texttt{ls}) will help you conquer the CLI.

\hypertarget{printing-your-working-directory}{%
\subsubsection{Printing your working
directory}\label{printing-your-working-directory}}

To know where you are you can see your current location, that is to
\emph{print your working directory} using the \texttt{pwd} command.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{pwd}
\end{Highlighting}
\end{Shaded}

\hypertarget{change-to-other-directory}{%
\subsubsection{Change to other
directory}\label{change-to-other-directory}}

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{cd}\NormalTok{ test{-}dir}
\end{Highlighting}
\end{Shaded}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]

Some basic arguments to navigate across your terminal:

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{cd}\NormalTok{ .. }\CommentTok{\# change backwards}
\BuiltInTok{cd}\NormalTok{ \textasciitilde{}  }\CommentTok{\# change to the home}
\BuiltInTok{cd}\NormalTok{ /  }\CommentTok{\# change to the root}
\BuiltInTok{cd} \AttributeTok{{-}}  \CommentTok{\# change to previous dir}
\end{Highlighting}
\end{Shaded}

\end{tcolorbox}

\hypertarget{listing-files}{%
\subsubsection{Listing files}\label{listing-files}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ls} 
\end{Highlighting}
\end{Shaded}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
You can navigate your executed commands by typing \faIcon{arrow-up} or
\faIcon{arrow-down}.
\end{tcolorbox}

\hypertarget{making-new-directories}{%
\subsubsection{Making new directories}\label{making-new-directories}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mkdir}\NormalTok{ test{-}dir}
\end{Highlighting}
\end{Shaded}

\hypertarget{creating-a-file}{%
\subsubsection{Creating a file}\label{creating-a-file}}

A simple command to create any file inside your terminal is
\texttt{touch} it just create a file, but do not allow any editing.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{touch}\NormalTok{ new{-}file.txt}
\end{Highlighting}
\end{Shaded}

The \texttt{new-file.txt} is empty and created on your current location
unless you assign another path when creating it. We suggest to take a
look at \href{https://github.com/allisonhorst/}{Allison Horst}
illustrations, especially on how to name files depending on the
\emph{case} see \textbf{?@fig-naming-files}

\hypertarget{printing-files-to-the-screen}{%
\subsubsection{Printing files to the
screen}\label{printing-files-to-the-screen}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{cat}\NormalTok{ new{-}file.txt}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{some}
\ExtensionTok{lines}
\ExtensionTok{that}
\ExtensionTok{were}
\ExtensionTok{written}
\end{Highlighting}
\end{Shaded}

\hypertarget{removing-files-or-directories}{%
\subsubsection{Removing files or
directories}\label{removing-files-or-directories}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rm}
\end{Highlighting}
\end{Shaded}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
When having a long command, it becomes practically to go to the
beginning or to the end of it. To do so you can use the key combination
\texttt{Ctrl\ +\ A} and \texttt{Ctrl\ +\ E} respectively.
\end{tcolorbox}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rmdir}
\end{Highlighting}
\end{Shaded}

\hypertarget{anatomy-of-a-command}{%
\subsection{Anatomy of a command}\label{anatomy-of-a-command}}

There is still many conventions by which the parts of a command line
might be called, yet a very standard convention is presented in
Fig.~\ref{fig-command}

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/command-anatomy.png}

}

\caption{\label{fig-command}A simple command and a convention to call
its main components}

\end{figure}

Some other for instance also tend to call the \texttt{option} as
\texttt{flag}. This conventions are powerful because almost any command
line interface display this structure (complex one add some other
features and simple one tend to lack subcommands).

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Challenge}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
Bacterial defense mechanisms to avoid bacteriophage infections are
abundant. One of these is the
\href{https://en.wikipedia.org/wiki/Restriction_modification_system}{:restriction-modification
system (RM-System)}, which works by targeting a specific sites called
\emph{motifs}, shared by the phage and bacteria, with methylations.
Motifs are commonly represented as a
\href{https://en.wikipedia.org/wiki/Sequence_logo}{:sequence logo} which
is a probabilistic representation of the nucleotides at each position.
The challenge consists of finding the number of times the motif from
Fig.~\ref{fig-motif} appears on \emph{B. tequilensis} EA-CB0015 genome
using a command. Assume that probabilities are equal when multiple bases
appeared at one site.

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/motif.png}

}

\caption{\label{fig-motif}A RM-system motif logo}

\end{figure}

Before diving into an \protect\hyperlink{motif-search}{:answer} take
your time to think and solve it by your own.
\end{tcolorbox}

\hypertarget{most-important-skills}{%
\section{Most important skills}\label{most-important-skills}}

When facing the CLI several issues or problems will arise. As for any
other unintuitive challenge, a complete text interface Handling errors.
Getting help Patience

\hypertarget{intermediate-unix-commands}{%
\section{Intermediate Unix commands}\label{intermediate-unix-commands}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{wc}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tr}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep} 
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sed} 
\end{Highlighting}
\end{Shaded}

\hypertarget{special-operators-or-metacharacters}{%
\section{Special operators or
metacharacters}\label{special-operators-or-metacharacters}}

Some operator or metacharacters have special functions in bash. For
instance the \texttt{*} or \emph{wildcard} is a regular expression
character (sometimes called as a placeholder) that will turn in
\emph{any character, many times}, similarly the \texttt{?} represents
\emph{any character, once}. Whereas the \texttt{\$} (dollar sign or
operator) is intended for an special task: call \emph{environmental
variables} which means that once a variable is defined (e.g.,
\texttt{var=1}) this variable can be called via the \texttt{\$} operator
anytime \texttt{echo\ \$var} will get us \texttt{1} as the standard
output

\hypertarget{unix-flows}{%
\section{Unix flows}\label{unix-flows}}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
When using the CLI at first its common to feal quite slow. Then, a very
useful tip to boost the productivity from the command line is the
autocompletion of commands by hitting
\texttt{\textless{}tab\textgreater{}} after the initial command.
\end{tcolorbox}

\hypertarget{redirection}{%
\subsection{Redirection}\label{redirection}}

\begin{figure}

{\centering \includegraphics[width=0.7\textwidth,height=\textheight]{chs-command-line/imgs-cli/redirection.png}

}

\caption{Redirecting flow}

\end{figure}

\hypertarget{pipe}{%
\subsection{Pipe}\label{pipe}}

\begin{figure}

{\centering \includegraphics[width=0.7\textwidth,height=\textheight]{chs-command-line/imgs-cli/pipeflow.png}

}

\caption{Pipe flow}

\end{figure}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
When having a long command, it is also useful to jump by lines instead
of character by character. To do so you can use the key combination
\texttt{Alt\ +\ \textless{}-} and \texttt{Alt\ +\ -\textgreater{}}
respectively.
\end{tcolorbox}

\hypertarget{loops-conditionals-and-script-variables}{%
\section{loops, conditionals and script
variables}\label{loops-conditionals-and-script-variables}}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Challenge}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
A second part of this challenges consists of create a script out from r
the motif-search one-line command that recursively search the motifs in
all genomes from a zip file that contains 10 bacterial genomes. The
script should include the shebang, loops, conditionals and environmental
variables.

See a possible script that solve the challenge
\protect\hyperlink{motif-search-in-multiple-files}{:here}
\end{tcolorbox}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Challenge}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
\href{https://en.wikipedia.org/wiki/Restriction_enzyme}{:Restriction
endonucleases} (RE) cleave the DNA by digesting the
\href{https://en.wikipedia.org/wiki/Phosphodiester_bond}{:phosphodiester
bond} between two nucleotides. Many RE are directed to specific DNA
motifs normally palindromic. There are mainly three types according to
it digestive mechanism. RE have been widely used in molecular
biotechnology because its specificity and versatility to carry out
different experiments.

One of the main uses of RE is to generate a pattern of restricted
fragments from different organisms so that samples of organisms,
sequences or genes could be distinguished, as long as they display
differences in the number of recognition motifs. This is normally done
in the lab, where an RE is mixed with a DNA sample and later an
\href{https://en.wikipedia.org/wiki/Electrophoresis}{:electrophoresis}
gel is run to see a separation pattern according to the fragments size.

Professor Javier has sequenced the genome of a sampled SARS-CoV2 and
want to see the band pattern that the genome would display if it were
digested with the RE EcoRV. He has asked you to help him with this
problem. The expected output is a text file with the sizes of the
fragments, where the size is the number of nucleotides of each fragment.
\end{tcolorbox}

For more explanations on the basic commands in the command line we
suggest to visit the first chapters of \emph{Computing skills for
biologist} from Allesina and Wilmes (2019)

\hypertarget{sec-vcs}{%
\chapter{Version control}\label{sec-vcs}}

\begin{quote}
Here we will first address what is version control, its importance and
basics on the local workflow.

Later we will introduce GitHub and explore more advanced commands for
the collaboration workflow
\end{quote}

\hypertarget{version-control-systems}{%
\section{Version control systems}\label{version-control-systems}}

As its name suggests, a version control system (VCS) allow you to keep
record of the changes happening while working files and directories.
Several VCSs have been created but the most popular is
\href{https://git-scm.com}{git}. It is characterized by being a
\textbf{distributed} VCS, which means that changes history are recorded
locally (whether in a user laptop or user account) in contrast to other
\textbf{centralized} VCSs that changes are saved on a shared machine or
server.

So, why bother to learn a VCS in bioinformatics? Well there are many
reasons, but to highlight some of them: i) Since VCSs allow you to
record changes, you can always trace back the steps made in ana
analysis, which is nice for the \textbf{reproducibility} of your work.
ii) a system like git could be coupled with a shared-centralized server
as it is \href{https://github.com}{GitHub} (we'll talk about it
\protect\hyperlink{exploring-github}{later} and then one could
\textbf{share and collaborate}, expanding the extent of your research
and iii) following the structures and command from git its at first
overwhelming and demands consistency and order, then when scaling a
project it will payoff this stepping curve of learning by keeping the
\textbf{efficiency} of your work.

\hypertarget{git-installation-and-configuration}{%
\section{Git installation and
configuration}\label{git-installation-and-configuration}}

Installation could proceed from the official page of
\href{https://git-scm.com}{git}. If working from WSL it has the binaries
preinstalled, so you can jump directly to the configuration. The second
step is to configure your user name and an email. with the following
lines:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{git}\NormalTok{ config }\AttributeTok{{-}{-}global}\NormalTok{ user.name }\StringTok{"Your Name"}
\FunctionTok{git}\NormalTok{ config }\AttributeTok{{-}{-}global}\NormalTok{ user.email user@eafit.edu.co}
\end{Highlighting}
\end{Shaded}

You could always user the preferred e-mail. More configurations are
available, for instance the preferred editor to work with and so on, you
can explore by asking for help \texttt{git\ config\ -\/-help} or
\texttt{git\ config\ -\/-list}.

\hypertarget{the-basics}{%
\subsection{The basics}\label{the-basics}}

There are at least \textbf{six} basic commands. Three of them allow
recording local changes (\texttt{git\ init}, \texttt{git\ add} and
\texttt{git\ commit}) and the other three help you to inspect the state
of the changes (\texttt{git\ status}, \texttt{git\ diff} and
\texttt{git\ log}), we will dive into the detail in the following lines.
So, to start recording changes in a directory you must
\textbf{initialize} the directory (which will now be called repository)
using \texttt{git\ init}. This is a one-time command to get started.

\hypertarget{the-local-workflow}{%
\subsection{The local workflow}\label{the-local-workflow}}

\begin{figure}

{\centering \includegraphics[width=0.6\textwidth,height=\textheight]{chs-command-line/imgs-cli/gitlocal.png}

}

\caption{Local workflow of a git project adding and committing changes}

\end{figure}

\#local-workflow

\hypertarget{exploring-github}{%
\section{Exploring GitHub}\label{exploring-github}}

As mention before, one of the great advantages that git can later
achieve is to collaborate. However, to get into that a sharable server
must allow users to have a common ground, and this is what
\href{https://github.com/home}{GitHub} allow. GitHub is a web platform
where the local repositories become public and any user can access to
the controlled versions of an image of your repo. The famous pet is the
octocat Fig.~\ref{fig-octocat}

\begin{marginfigure}

{\centering \includegraphics{chs-command-line/imgs-cli/octocat.png}

}

\caption{\label{fig-octocat}The octocat from GitHub}

\end{marginfigure}

To enable the communication with a remote repo, git has encoded many
specific commands, once the repos are \textbf{cloned} a simple workflow
from the own local and remote repos is made possible thanks to two
simple commands \texttt{git\ push} and \texttt{git\ pull}.

There a several ways to starting out a remote project, whether it starts
from a local folder or whether it starts from a remote repo. The second
strategy is sometimes easier as you just need to later
\texttt{git\ clone} the remote into a local folder. To do so every repo
has its own code-icon \faIcon{code} to later copy the repo link and
later hit \texttt{git\ clone\ \textless{}https...\textgreater{}} on the
desired folder. Now you got a linked copy of the remote on your machine.

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-caution-color}{\faFire}\hspace{0.5em}{Caution}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-caution-color-frame, colbacktitle=quarto-callout-caution-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
But before working on a remote and pushing your first commits, it is
common to find an error regarding the remote branch (also called
\emph{origin}). There are several ways to avoid this caveat, but a very
anticipated way is to configure git

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{git}\NormalTok{ config }\AttributeTok{{-}{-}global}\NormalTok{ push.autoSetupRemote true}
\end{Highlighting}
\end{Shaded}

This will save you from every time typing
\texttt{git\ push\ -\/-set-upstream\ origin\ \textless{}main\textgreater{}}
when working on a new remote repo.
\end{tcolorbox}

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/gitcloning.png}

}

\caption{\label{fig-cloning}The cloning workflow in brief}

\end{figure}

Cloning is therefore the process of creating a local copy of remote
repository, that is a machine version of the remote repo, later all the
common local workflow is carried out normally Fig.~\ref{fig-cloning}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-tip-color}{\faLightbulb}\hspace{0.5em}{Tip}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-tip-color-frame, colbacktitle=quarto-callout-tip-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
If you want to keep a file out of synchronization git provides a simple
way to do so by creating a \texttt{.gitignore} file having the the paths
to the files to keep in your local machine
\end{tcolorbox}

\hypertarget{forking-and-collaborate}{%
\subsection{Forking and collaborate}\label{forking-and-collaborate}}

Basic collaboration on an open repository is a three-step process. First
you need to \texttt{fork} the repo, this will create a mirror copy of
the repo in your GitHub profile. In a second step, a simple
\texttt{clone} of the repo will generate a copy of the forked repo on
your local machine, so you can freely work and make your mistakes and
push them to your forked repo that belongs to your account. In a third
step, once they are on the remote repo you will have to create a
\texttt{pull\ request} (PR), as its name suggest: you are
\textbf{asking} the owners of the original to consider your changes
Fig.~\ref{fig-forking}.

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/gitforking.png}

}

\caption{\label{fig-forking}A common collaborative workflow from GitHub,
using fork, clone and pull requests}

\end{figure}

\hypertarget{branching-and-merging}{%
\subsection{Branching and merging}\label{branching-and-merging}}

\hypertarget{sec-managers}{%
\chapter{Package managers}\label{sec-managers}}

\hypertarget{what-is-the-importance-of-package-managers}{%
\section{What is the importance of package
managers}\label{what-is-the-importance-of-package-managers}}

\hypertarget{conda-environments-and-other-package-managers}{%
\section{Conda environments and other package
managers}\label{conda-environments-and-other-package-managers}}

\hypertarget{creating-environments-with-conda}{%
\section{Creating environments with
Conda}\label{creating-environments-with-conda}}

\hypertarget{package-managers-for-os}{%
\section{Package managers for OS}\label{package-managers-for-os}}

There are several package managers handling general purpose packages and
apps. For MacOS the famous one is \href{}{Homebrew} and for Windows
several could be used such as \href{}{Chocolatey} and
\href{https://scoop.sh/}{Scoop}.

\hypertarget{sec-hpc}{%
\chapter{Notions of HPC}\label{sec-hpc}}

\hypertarget{serial-computing}{%
\section{Serial computing}\label{serial-computing}}

\begin{figure}

{\centering \includegraphics{chs-command-line/imgs-cli/serial-computing.png}

}

\caption{serial computing}

\end{figure}

\hypertarget{parallel-computing}{%
\section{Parallel computing}\label{parallel-computing}}

\includegraphics{chs-command-line/imgs-cli/parallel-computing.png}

\hypertarget{computer-architectures}{%
\section{Computer architectures}\label{computer-architectures}}

\begin{marginfigure}

{\centering \includegraphics[width=0.7\textwidth,height=\textheight]{chs-command-line/imgs-cli/laptop-arch.png}

}

\caption{Laptop processor architecture}

\end{marginfigure}

\begin{figure}

{\centering \includegraphics[width=0.6\textwidth,height=\textheight]{chs-command-line/imgs-cli/partition-arch.png}

}

\caption{HPC architecture}

\end{figure}

\hypertarget{using-apolo-cluster-with-slurm}{%
\section{Using Apolo cluster with
Slurm}\label{using-apolo-cluster-with-slurm}}

\includegraphics{chs-command-line/imgs-cli/cluster-process.png}

Slurm is a job manager scheduler. It brings order to the jobs sent by
users of the clusters and facilitate the process of check the state of
the job in the cluster and the actual state of the cluster itself. To do
so it has basically three main commands. The first one is \texttt{sinfo}

\part{Sequence analysis}

\hypertarget{sec-seq-analysis}{%
\chapter{Sequence analysis}\label{sec-seq-analysis}}

\begin{quote}
In this chapter we we will consider several biological concepts that
appear central to understand the manipulation of biological data.

We are also covering the transition of a biological (organic)
information to digital bits.

We will explore some databases in which biological information is
stored.
\end{quote}

\hypertarget{biological-information}{%
\section{Biological information}\label{biological-information}}

Since the origin, organisms (or molecules) are the result of different
selective processes. An emergent property of successive iterations of
survival/decease was the ability of molecules to keep a record of its
past in a very stable manner, so that it will pass generation to
generation. Although this might not be the first property or molecule to
ever exist (see the
\href{https://en.wikipedia.org/wiki/Abiogenesis}{:origin of life}), the
innovation of organisms to pack information of previous events became so
advantageous that almost any form of living organism has this property:
that is DNA Fig.~\ref{fig-chromosome}.

DNA

\begin{figure}

{\centering \includegraphics{chs-sequence-analysis/imgs-seqan/chromosome-genes.png}

}

\caption{\label{fig-chromosome}The different scales of information
packed in a biological fashion. Figure adapted from Meneely et al.
(2017) book showing how \emph{DNA in a eukaryote is packaged with
chromosomal proteins to form the much more compact chromatin. Chromatin
is then folded into higher-order structures and eventually into the
chromosome}}

\end{figure}

\hypertarget{the-duality-of-dna}{%
\section{The duality of DNA}\label{the-duality-of-dna}}

DNA is an organic molecule build upon several more organic bricks named
nucleotides arranged in a very compacted helix structure and too much
can be understood from its molecular nature. It is of course governed by
the physicochemical properties of the atoms, therefore DNA is a
physicochemical entity. But those common bricks have also a very
beautiful emergent property that comes out from their order (mainly),
and that is to keep the instructions that build the organism from which
it belongs. These instructions are nothing else than information, which
also follow some informational rules. Then DNA is also an informational
entity.

This idea of the duality of the DNA (and of course of other biological
molecules) is the source of the study of many bioinformatic fields
\textbf{?@fig-biofields}.

\includegraphics{chs-sequence-analysis/.pdf}

\hypertarget{the-central-dogma-theory-of-molecular-biology-extended}{%
\section{\texorpdfstring{The central \sout{dogma} theory of molecular
biology
extended}{The central dogma theory of molecular biology extended}}\label{the-central-dogma-theory-of-molecular-biology-extended}}

\hypertarget{sequencing-strategies}{%
\section{Sequencing strategies}\label{sequencing-strategies}}

\hypertarget{sequencing-over-time}{%
\section{Sequencing over time}\label{sequencing-over-time}}

\hypertarget{some-insights-from-sequencing-genomes}{%
\section{Some insights from sequencing
genomes}\label{some-insights-from-sequencing-genomes}}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Challenge}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
How many complete genomes of Bacillus subtilis are there in the NCBI
databases?

Count the number of features (genes, CDS, ncRNA, rRNA, etc.) in the
genome of Bacillus subtilis NCIB 3610 (GCF\_002055965.1)

Document each of the steps and how did you end up with the answer.
Saving the file with your initials (e.g.,
CG-activity01.\{md,txt,docx\})z
\end{tcolorbox}

\hypertarget{biological-information-databases}{%
\section{Biological information
databases}\label{biological-information-databases}}

\hypertarget{sec-sanger}{%
\chapter{Sanger analysis}\label{sec-sanger}}

\begin{quote}
This is a section about one of the first gen sequencing tech, that is
Sanger sequencing and its principles and evolution, strengths and
limitations

Also will be covering the bioinformatics side of the method: what kind
of files we get out Sanger, how we process them and an important
paradigm that came out from this sequence method.
\end{quote}

\hypertarget{sanger-sequencing-methods}{%
\section{Sanger sequencing methods}\label{sanger-sequencing-methods}}

\hypertarget{the-chain-termination-method}{%
\subsection{The chain termination
method}\label{the-chain-termination-method}}

\hypertarget{sanger-with-capillary-electrophoresis}{%
\subsection{Sanger with capillary
electrophoresis}\label{sanger-with-capillary-electrophoresis}}

\hypertarget{strengths-and-limitations-of-sanger-methods}{%
\subsection{Strengths and limitations of Sanger
methods}\label{strengths-and-limitations-of-sanger-methods}}

\hypertarget{files-from-sanger-sequencing}{%
\section{Files from Sanger
sequencing}\label{files-from-sanger-sequencing}}

\hypertarget{sanger-processing-workflow}{%
\section{Sanger processing workflow}\label{sanger-processing-workflow}}

\hypertarget{the-16s-rrna-and-its-relevance-for-sequencing}{%
\section{The 16S rRNA and its relevance for
sequencing}\label{the-16s-rrna-and-its-relevance-for-sequencing}}

\hypertarget{sec-aliignments}{%
\chapter{Sequence alignments}\label{sec-aliignments}}

\hypertarget{why-do-we-align-sequences}{%
\section{Why do we align sequences?}\label{why-do-we-align-sequences}}

In search of homology and identy

\hypertarget{what-is-homology}{%
\section{What is homology}\label{what-is-homology}}

\hypertarget{pairwise-alignments-algorihtms}{%
\section{Pairwise alignments
algorihtms}\label{pairwise-alignments-algorihtms}}

\hypertarget{hamming-distance}{%
\subsection{Hamming distance}\label{hamming-distance}}

\hypertarget{edit-distance}{%
\subsection{Edit distance}\label{edit-distance}}

\hypertarget{dynamic-programming}{%
\subsubsection{Dynamic programming}\label{dynamic-programming}}

\hypertarget{needleman-wunsch-global-alignment}{%
\subsection{Needleman-Wunsch (global
alignment)}\label{needleman-wunsch-global-alignment}}

\hypertarget{smith-waterman-local-alignment}{%
\subsection{Smith-Waterman (local
alignment)}\label{smith-waterman-local-alignment}}

\hypertarget{the-genetic-code-and-scoring-matrices}{%
\section{The genetic code and Scoring
matrices}\label{the-genetic-code-and-scoring-matrices}}

\hypertarget{blast-and-its-families}{%
\section{BLAST and its families}\label{blast-and-its-families}}

psi-blast? true homologs, recurrent blast to polish scoring matrix
during several generations to generate true homologs

\hypertarget{multiple-sequence-alignments}{%
\section{Multiple sequence
alignments}\label{multiple-sequence-alignments}}

\hypertarget{sec-phylogenetics}{%
\chapter{Phylogenetics}\label{sec-phylogenetics}}

\hypertarget{what-is-a-phylogenetic-tree}{%
\section{What is a phylogenetic
tree}\label{what-is-a-phylogenetic-tree}}

\hypertarget{mehtods-for-phylogenetic-reconstruction}{%
\section{Mehtods for phylogenetic
reconstruction}\label{mehtods-for-phylogenetic-reconstruction}}

\hypertarget{building-a-phylogenetic-reconstruction}{%
\section{Building a phylogenetic
reconstruction}\label{building-a-phylogenetic-reconstruction}}

\hypertarget{evolutionary-substitution-model}{%
\subsection{Evolutionary substitution
model}\label{evolutionary-substitution-model}}

\hypertarget{maximum-likelihood}{%
\subsection{Maximum likelihood}\label{maximum-likelihood}}

\hypertarget{bayesian-inference}{%
\subsection{Bayesian inference}\label{bayesian-inference}}

\hypertarget{sec-ngs-tgs}{%
\chapter{NGS and TGS: principles}\label{sec-ngs-tgs}}

\hypertarget{platforms-yields}{%
\section{Platforms yields}\label{platforms-yields}}

\hypertarget{reads-main-differences}{%
\section{Reads main differences}\label{reads-main-differences}}

\hypertarget{illumina-principle-sequencing-by-synthesis}{%
\section{Illumina principle (sequencing by
synthesis)}\label{illumina-principle-sequencing-by-synthesis}}

\hypertarget{the-fastq-format}{%
\subsection{The fastq format}\label{the-fastq-format}}

\hypertarget{quality-assesment-of-illumina}{%
\subsection{Quality assesment of
Illumina}\label{quality-assesment-of-illumina}}

\hypertarget{pacbio-principle-sequencing-by-incorporation}{%
\section{PacBio principle (sequencing by
incorporation)}\label{pacbio-principle-sequencing-by-incorporation}}

\hypertarget{throughput-evolution}{%
\subsection{Throughput evolution}\label{throughput-evolution}}

\hypertarget{quality-assesment-of-pacbio}{%
\subsection{Quality assesment of
PacBio}\label{quality-assesment-of-pacbio}}

\hypertarget{oxford-nanopore-technology-ont-principle}{%
\section{Oxford Nanopore Technology (ONT)
principle}\label{oxford-nanopore-technology-ont-principle}}

\hypertarget{platforms}{%
\subsection{Platforms}\label{platforms}}

\hypertarget{the-fast5-file-format}{%
\subsection{The fast5 file format}\label{the-fast5-file-format}}

\part{Genomics}

\hypertarget{sec-genomics}{%
\chapter{Genome assembly}\label{sec-genomics}}

\hypertarget{the-problem-of-assembling-genomes}{%
\section{The problem of assembling
genomes}\label{the-problem-of-assembling-genomes}}

\hypertarget{main-algorithms-for-genome-asssembly}{%
\section{Main algorithms for genome
asssembly}\label{main-algorithms-for-genome-asssembly}}

\hypertarget{overlay-layout-consensus-ols}{%
\subsection{Overlay, Layout, Consensus
(OLS)}\label{overlay-layout-consensus-ols}}

\hypertarget{de-bruijn-graphs}{%
\subsection{De Bruijn graphs}\label{de-bruijn-graphs}}

\hypertarget{main-concepts-of-an-assembly}{%
\section{Main concepts of an
assembly}\label{main-concepts-of-an-assembly}}

\hypertarget{contigs-unitigs-scaffolds}{%
\subsection{Contigs, Unitigs,
Scaffolds}\label{contigs-unitigs-scaffolds}}

\hypertarget{a-complete-workflow-for-assembling-genomes}{%
\section{A complete workflow for assembling
genomes}\label{a-complete-workflow-for-assembling-genomes}}

\hypertarget{assessing-genomes}{%
\section{Assessing genomes}\label{assessing-genomes}}

\hypertarget{inspecting-genome-graphs}{%
\subsection{Inspecting genome graphs}\label{inspecting-genome-graphs}}

\hypertarget{genome-completeness}{%
\subsection{Genome completeness}\label{genome-completeness}}

\hypertarget{understanding-genome-difficulties}{%
\section{Understanding genome
difficulties}\label{understanding-genome-difficulties}}

\begin{itemize}
\item
  End of chromosomes
\item
  Erros
\item
  Lack of coverage
\item
  Heterozigozity
\item
  repeats
\end{itemize}

\hypertarget{sec-genome-annotation}{%
\chapter{Genome annotation}\label{sec-genome-annotation}}

\hypertarget{ab-initio-annotation}{%
\section{\texorpdfstring{\emph{ab initio}
annotation}{ab initio annotation}}\label{ab-initio-annotation}}

\hypertarget{homology-annotation}{%
\section{Homology annotation}\label{homology-annotation}}

\hypertarget{annotation-files}{%
\section{Annotation files}\label{annotation-files}}

\hypertarget{the-gbk-and-gbff}{%
\subsection{the GBK and GBFF}\label{the-gbk-and-gbff}}

\hypertarget{the-gff-specifications}{%
\subsection{The GFF specifications}\label{the-gff-specifications}}

\hypertarget{visualizing-genomes-and-annotations}{%
\section{Visualizing genomes and
annotations}\label{visualizing-genomes-and-annotations}}

\hypertarget{sec-variant-calling}{%
\chapter{Variant calling analysis}\label{sec-variant-calling}}

\hypertarget{common-mutations}{%
\section{Common mutations}\label{common-mutations}}

\hypertarget{structural-variants}{%
\section{Structural variants}\label{structural-variants}}

\hypertarget{genome-rearrangements}{%
\section{Genome rearrangements}\label{genome-rearrangements}}

\hypertarget{read-mapping-algorithms-and-programs}{%
\section{Read mapping algorithms and
programs}\label{read-mapping-algorithms-and-programs}}

\hypertarget{burrow-wheeler-alignment}{%
\subsection{Burrow-Wheeler-Alignment}\label{burrow-wheeler-alignment}}

\hypertarget{bwa-mem2}{%
\subsection{BWA-MEM2}\label{bwa-mem2}}

\hypertarget{minimap2}{%
\subsection{Minimap2}\label{minimap2}}

\hypertarget{sam-bam-and-cram-formats}{%
\subsection{SAM, BAM and CRAM formats}\label{sam-bam-and-cram-formats}}

\hypertarget{identifying-mutations}{%
\section{Identifying mutations}\label{identifying-mutations}}

\hypertarget{freebayes-and-snippy}{%
\subsection{Freebayes and Snippy}\label{freebayes-and-snippy}}

\hypertarget{the-vcf-file}{%
\subsection{The VCF file}\label{the-vcf-file}}

\part{Structural bioinformatics}

\hypertarget{sec-biostructural}{%
\chapter{Structural bioinformatics}\label{sec-biostructural}}

Structural bioinformatics is a multidisciplinary area enriched by
chemistry, physics, computer science and many others. Although, it could
be focused on different biological macro-molecules, here the emphasis
will be focused on proteins.

One of the first protein structure elucidated was myoglobin and it
triggered the study of the role of the structure of proteins and its
biological functions

Identify a protein related to your study that could be further analyzed.

\hypertarget{protein-structures}{%
\section{Protein structures}\label{protein-structures}}

Difference between the levels of protein structure primary structure is
the basic linear representation of aminoacids. Natural aminoacids and
modified or rare aminoacids display physico-chemical rich information
and could be represented by letters. Therefore in the genetic code we
could find the one-letter code.

Secondary structures result from the spatial arrangement of aminoacids
that interact with each closer neighbors. There are some remarkable
secondary structures such as \(\beta\)-sheets, \(\alpha\)-helix, coils
(flexible) and others.

Tertiary structure informs about the structural disposition of the
secondary structures that fold between each other due to hydrophobic
interactions, disulfure bonds, and other chemical interactions forming a
globular and dynamic structure. Thus, proteins could display multiple
structural states depending on the physical and energy stability (see
the
\href{https://en.wikipedia.org/wiki/Levinthal\%27s_paradox}{:Levinthal's
paradox}).

Quaternary structures result from interaction of multiple tertiary
structures. \emph{The structure, therefore dictates the protein
function}. This basic concept have triggered more recently a boom on the
analysis of the structure of proteins.

\hypertarget{identifying-or-predicting-protein-structures}{%
\section{Identifying or predicting protein
structures}\label{identifying-or-predicting-protein-structures}}

Xray crystallography, nuclear magnetic resonance (NMR) allows to
encapsulated dynamic information of the protein in time Electron
micrography (EM) and Cryo-EM. These experiment rely on highly
specialized set ups and there are other drawbacks.

{\marginnote{\begin{footnotesize}To date, helium is scarce around the
world, so labs all around are having trouble to get this
element.\end{footnotesize}}}

Modelling the structure whether \emph{ab initio} or by \emph{homology}
also allow structure prediction. However these strategies

Recently AlphaFold

The protein database (PDB)

Protein topologies resulting from the folding: horshoe, beta-barrel and
other could be identified

Structural classification of proteins (SCOP) when analyzing a new
protein classification by class, architecture, topology (fold-family),
homologous superfamily and sequence family

Importance of Gene Ontology

\hypertarget{secondary-structure-prediction}{%
\subsection{Secondary structure
prediction}\label{secondary-structure-prediction}}

Secondary structures could also be represented in one letter (e.g.)

Functional domains could be predicted by sequence alignment and allow
structural inference. Main predictions are based heavily on machine
learning and are frequently accepted under a consensus of multiple
tools.

{\marginnote{\begin{footnotesize}There is a group of protein called
intrinsically-disordered proteins that change in its tertiary structures
quite frequently, therefore it prediction could be
troublesome.\end{footnotesize}}}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, leftrule=.75mm, breakable, titlerule=0mm, toprule=.15mm, rightrule=.15mm, title=\textcolor{quarto-callout-warning-color}{\faExclamationTriangle}\hspace{0.5em}{Exercise 01}, colback=white, opacitybacktitle=0.6, toptitle=1mm, bottomtitle=1mm, colframe=quarto-callout-warning-color-frame, colbacktitle=quarto-callout-warning-color!10!white, arc=.35mm, coltitle=black, bottomrule=.15mm, left=2mm]
Submit 5lWM (JAK3) from the PDB on FASTA format on the
\href{http://www.compbio.dundee.ac.uk/jpred/}{JPRED} and
\href{http://bioinf.cs.ucl.ac.uk/psipred/}{PSIPRED} and compared with
the \href{https://www.rcsb.org/sequence/5LWM}{experimentally predicted}
version of the protein. Analyze the predictions and tell are there
differences between predictions? Which one is more accurate?
\end{tcolorbox}

\hypertarget{pdb-database-introduction}{%
\section{PDB database introduction}\label{pdb-database-introduction}}

The PDB database is one of the most important and ancient open
biological database where all new protein structures are submitted. It
is an international consortium where several regions work together to
curate information.

The PDB in Europe (\href{pdbe.org}{PDBe}) is not only for proteins but
form many other \emph{experimentally predicted} macro-molecules
(protein-protein interactions, peptides, RNA and so on).

Protein structures are registered using a unique code of four
characters.

X-ray crystallography: is a chemical state of the macro-molecule where
it is immobilized, therefore information correspond to one state of the
structure, then crystal protein is submitted to an x-ray beam to
generate a diffraction pattern.

NMR spectroscopy: captures dynamic information of the protein, but
generally resolves small proteins. The principle?

Electron microscopy adapted to cryo-preservation allow proteins
visualization.

\part{Challenges}

\hypertarget{unix-challenges}{%
\chapter*{Unix challenges}\label{unix-challenges}}
\addcontentsline{toc}{chapter}{Unix challenges}

\hypertarget{motif-search}{%
\section*{Motif search}\label{motif-search}}
\addcontentsline{toc}{section}{Motif search}

The first thing we must do to solve the
\protect\hyperlink{anatomy-of-a-command}{Challenge} is download the
sequence of interest. In \protect\hyperlink{anatomy-of-a-command}{Ch 1}
the genome of \emph{Bacillus tequilensis} EA-CB0015 is proposed. There
are several ways to do so. A simple way is to search in the NCBI
databases for the entire genome and download as \texttt{FASTA} file (we
cover the properties of this file format in
\protect\hyperlink{sec-seq-analysis}{Ch 5}).

When inspecting the file we see the following lines at the top

\begin{Shaded}
\begin{Highlighting}[]
\OperatorTok{\textgreater{}}\NormalTok{NZ\_CP048852.1 }\ExtensionTok{Bacillus}\NormalTok{ tequilensis strain EA{-}CB0015 chromosome, complete genome}
\ExtensionTok{ATGGAAAATATATTAGACCTGTGGAATCAAGCCCTTGCTCAAATCGAAAAAAAGTTGAGCAAACCGAGTTTTGAGACTTG}
\ExtensionTok{GATGAAATCCACCAAAGCCCACTCACTGCAGGGAGATACACTGACGATCACGGCTCCCAATGAATTTGCCAGAGACTGGC}
\ExtensionTok{TGGAGTCCAGATACTTGCATCTGATTGCAGATACTATATATGAATTAACCGGGGAAGAATTGAGCATTAAGTTTGTCATT}
\ExtensionTok{CCTCAAAATCAAGATGTTGAGGACTTTATGCCAAAGCCGCAAGTCAAAAAAGCGGTTAAAGAAGATACATCTGATTTTCC}
\end{Highlighting}
\end{Shaded}

The file is about 4 Mb and it is heavy to be opened in any editor.
Therefore when looking for a pattern it becomes more useful the CLI.

A first attempt to do is simply by using \texttt{grep} for the pattern
presented on Fig.~\ref{fig-motif}. So a first problem you will find is
to represent multiple characters in one position of the pattern. There
is a fantastic tool to so and its called \emph{regular expressions}
(regex). a regex is a way to encode groups of characters using single
characters. For instance the character \texttt{*} (wildcards) represents
any character, so it groups all characters and it is used broadly on the
CLI. There are many regex dialects however and for \texttt{grep} the dot
\texttt{.} represents any character and when the pipe
\texttt{\textbar{}} represents an special operator in the CLI, for grep
regex it represents and \texttt{OR} operator.

So, the second step is to create our regular expression. Note the there
are six characters in the expression that can adopt different
characters. So our first regex is
\texttt{TAAA(T\textbar{}A)(.)(T\textbar{}A)(G\textbar{}A)(.)(C\textbar{}A)GGT}.
In the expression the parenthesis allow groups of characters.

Now lets see if we can search for that pattern using \texttt{grep} and
the appropriate option for using regex, which in this case is
\texttt{-E}. Another simply way to use the \texttt{grep\ -E} combination
is to use simple \texttt{egrep} command plus the \texttt{-c} option to
count the number of \emph{lines} where the pattern appears.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{egrep} \AttributeTok{{-}c} \StringTok{"TAAA(T|A)(.)(T|A)(G|A)(.)(C|A)GGT"}\NormalTok{ data/Bteq{-}genome.fasta}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
43
\end{verbatim}

Is this the solution? One might think it is solved. However, there are
several issues to be aware. First \texttt{egrep\ -c} is counting the
lines, what if in one line there are multiple matches? Then, the
approach will showing a value under the actual number of motifs. So a
way to solve that is to capture and extract the matches and print them
out. A way to do that is to use the option \texttt{-o} which actually
prints only the matching part of the lines, and then count them.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{egrep} \AttributeTok{{-}oc} \StringTok{"TAAA(T|A)(.)(T|A)(G|A)(.)(C|A)GGT"}\NormalTok{ data/Bteq{-}genome.fasta}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
43
\end{verbatim}

The outcome is the same tough. So is it solved? Every NCBI genome that
is downloaded display a distinct character besides de nucleotide
sequence. Every several number of nucleotides there is a new line
\texttt{\textbackslash{}n} that enables the wrapping file into multiple
lines. Therefore some motifs could be shopped by the
\texttt{\textbackslash{}n} character so they will never appear with the
regex used. A way to circumvent this issue is deleting the
\texttt{\textbackslash{}n} in the entire file. We can easily do that
with the translate \texttt{tr} command:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tr} \AttributeTok{{-}d} \StringTok{\textquotesingle{}\textbackslash{}n\textquotesingle{}} \OperatorTok{\textless{}}\NormalTok{ data/Bteq{-}genome.fasta}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\OperatorTok{\textgreater{}}\NormalTok{NZ\_CP048852.1 }\ExtensionTok{Bacillus}\NormalTok{ tequilensis strain EA{-}CB0015 chromosome, complete genomeATGGAAAATATATTAGACCTGTGGAATCAAGCCCTTGCTCAAATCGAAAAAAAGTTGAGCAAACCGAGTTTTGAGACTTGGATGAAATCCACCAAAGCCCACTCACTGCAGGGAGATACACTGACGATCACGGCTCCCAATGAATTTGCCAGAGACTGGCTGGAGTCCAGATACTTGCATCTGATTGCAGATACTATATATGAATTAACCGGGGAAGAATTGAGCATTAAGTTTGTCATTCCTCAAAATCAAGATGTTGAGGACTTTATGCCAAAGCCGCAAGTCAAAAAAGCGGTTAAAGAAGATACATCTGATTTTCC...}
\end{Highlighting}
\end{Shaded}

The genome now lies in an entire and unique line along with its title. A
way to chain the commands without creating intermediate files is by
using the pipe \texttt{\textbar{}} character. At the end we will use
\texttt{wc\ -l} to count the occurrence (although \texttt{grep} has its
own count option \texttt{\textbackslash{}c} it is inconvenient because
it will count the lines and now the lines of the exact matches provided
by the \texttt{-o} option and as we have just one line, then only will
count to 1).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tr} \AttributeTok{{-}d} \StringTok{\textquotesingle{}\textbackslash{}n\textquotesingle{}} \OperatorTok{\textless{}}\NormalTok{ data/Bteq{-}genome.fasta }\KeywordTok{|}
    \FunctionTok{egrep} \AttributeTok{{-}o} \StringTok{"TAAA(T|A)(.)(T|A)(G|A)(.)(C|A)GGT"} \KeywordTok{|}
    \FunctionTok{wc} \AttributeTok{{-}l}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
      50
\end{verbatim}

Are there yet? Well it turns out that the genome file contains only the
5'-3' strand of the genome. What about the other strands patterns? To
search them we need to generate the reverse-complement (RC) of the motif
so we are sure that they will appear when looking the 5'-3' strand. The
RC regex will be
\texttt{ACC(T\textbar{}G)(.)(T\textbar{}C)(T\textbar{}A)(.)(T\textbar{}A)TTTA"}
and adding this regex to the search willl be:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tr} \AttributeTok{{-}d} \StringTok{\textquotesingle{}\textbackslash{}n\textquotesingle{}} \OperatorTok{\textless{}}\NormalTok{ data/Bteq{-}genome.fasta }\KeywordTok{|}
    \FunctionTok{egrep} \AttributeTok{{-}o} \StringTok{"TAAA(T|A)(.)(T|A)(G|A)(.)(C|A)GGT|ACC(T|G)(.)(T|C)(T|A)(.)(T|A)TTTA"} \KeywordTok{|}
    \FunctionTok{wc} \AttributeTok{{-}l}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
      94
\end{verbatim}

\hypertarget{motif-search-in-multiple-files}{%
\section*{Motif search in multiple
files}\label{motif-search-in-multiple-files}}
\addcontentsline{toc}{section}{Motif search in multiple files}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\#! /usr/bin/env bash}

\VariableTok{motif}\OperatorTok{=}\StringTok{"TAAA(T|A)(.)(T|A)(G|A)(.)(C|A)GGT|ACC(T|G)(.)(T|C)(T|A)(.)(T|A)TTTA"}

\ControlFlowTok{for}\NormalTok{ f }\KeywordTok{in} \VariableTok{$(}\FunctionTok{ls} \VariableTok{$1)}\KeywordTok{;} \ControlFlowTok{do}
  \ControlFlowTok{if} \BuiltInTok{[} \OtherTok{{-}s} \VariableTok{$\{1\}}\NormalTok{/}\VariableTok{$\{f\}} \BuiltInTok{]}\KeywordTok{;} \ControlFlowTok{then}
      \VariableTok{motifnumber}\OperatorTok{=}\VariableTok{$(}\FunctionTok{tr} \AttributeTok{{-}d} \StringTok{\textquotesingle{}\textbackslash{}n\textquotesingle{}} \OperatorTok{\textless{}} \VariableTok{$\{1\}}\NormalTok{/}\VariableTok{$\{f\}} \KeywordTok{|} \FunctionTok{egrep} \AttributeTok{{-}o} \VariableTok{$motif} \KeywordTok{|} \FunctionTok{wc} \AttributeTok{{-}l}\VariableTok{)}
      \VariableTok{header}\OperatorTok{=}\VariableTok{$(}\FunctionTok{head} \AttributeTok{{-}n}\NormalTok{ 1 }\VariableTok{$\{1\}}\NormalTok{/}\VariableTok{$\{f\})}
      \BuiltInTok{echo} \AttributeTok{{-}e} \StringTok{"There are }\VariableTok{$motifnumber}\StringTok{ motifs in }\VariableTok{$header}\StringTok{"}
  \ControlFlowTok{else}
      \BuiltInTok{echo} \StringTok{"}\VariableTok{$f}\StringTok{ is empty"}
  \ControlFlowTok{fi}
\ControlFlowTok{done}
\end{Highlighting}
\end{Shaded}

To execute the command make sure to pass a folder with the decompressed
genomes as the first argument.

\hypertarget{fragments-digestion}{%
\section*{Fragments digestion}\label{fragments-digestion}}
\addcontentsline{toc}{section}{Fragments digestion}

\bookmarksetup{startatroot}

\hypertarget{references}{%
\chapter*{References}\label{references}}
\addcontentsline{toc}{chapter}{References}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-allesina2019}{}}%
Allesina, Stefano, and Madlen Wilmes. 2019. {``Computing Skills for
Biologists,''} January. \url{https://doi.org/10.2307/j.ctvc77jrc}.

\leavevmode\vadjust pre{\hypertarget{ref-meneely2017}{}}%
Meneely, Philip Mark, Rachel Dawes Hoang, Iruka N Okeke, and Katherine
Heston. 2017. \emph{Genetics: Genes, Genomes, and Evolution}. Oxford
University Press.

\end{CSLReferences}



\end{document}
