---
execute: 
  engine: knitr
filters: 
  - lightbox
  - nutshell
  - code-filename
lightbox: auto
---

# Sanger analysis {#sec-sanger}

> This is a section about perhaps one of the first generation sequencing technology, that is Sanger sequencing and its principles and evolution, strengths and limitations
>
> This chapter will also cover the bioinformatics side of the method: what kind of files we get out Sanger, how we process them and an important paradigm that came out from this sequence method.
>
> Finally we will dive into a study case that its very common in microbiology: the use of the 16S rRNA gene for taxonomy identification

## The first sequencing methods

Back in 70s there two different methods aiming to determine the base sequences of DNA and they basically operated at the level of  base per base reconstruction of the information. The most used during those days was so called [:Maxam-Gilbert](https://en.wikipedia.org/wiki/Maxam–Gilbert_sequencing) method and the second one was precisely the chain termination method or dideoxy sequencing or more commonly nowadays [:Sanger sequencing](https://en.wikipedia.org/wiki/Sanger_sequencing). Both methods named after their creators. It was, however the Sanger method the one that gradually became the facto method and continue to be the standard for the next thirty years. In fact, assembled genomes before the 2000s were sequenced with Sanger method, which included the first draft of the Human genome and many other organism @brown2018.

![A globular structure of a bacterial DNA polymerase](imgs-seqan/dna-pol.png){#fig-dna-pol .column-margin}

### The chain termination method

To under stand the chain termination method is important have a clear view of the DNA structure and chemical reaction that enable its polymerization. Whe the 

![The deoxy-ribose and dideoxy-ribose and consequences on the polymerization of a DNA strand](imgs-seqan/dideoxy-reaction.png){#fig-dideoxy-reaction}

:::{#fig-yt-sanger .column-margin}
{{< video https://www.youtube.com/embed/Wgv9QhjLTC8 >}}

A step-by-step of the Sanger sequencing method or the chain termination method.
:::

### Sanger with capillary electrophoresis

### Strengths and limitations of Sanger methods

## Files from Sanger sequencing

Several files are generated from a modern Sanger sequencing project. 

::: {.callout}
## {{< fa file >}} PHD.1

```
BEGIN_SEQUENCE H200824-010_A11_455-WT_spo0B_spo0B-R

BEGIN_COMMENT

CHROMAT_FILE: H200824-010_A11_455-WT_spo0B_spo0B-R
BASECALLER_VERSION: KB 1.4.0
TRACE_PROCESSOR_VERSION: KB 1.4.0
QUALITY_LEVELS: 99
TIME: Mon Aug 24 22:05:23 2020
TRACE_ARRAY_MIN_INDEX: 0
TRACE_ARRAY_MAX_INDEX: 18141
TRIM: -1 -1 -1.000000e+000
TRACE_PEAK_AREA_RATIO: -1.000000e+000
CHEM: term
DYE: big

END_COMMENT

BEGIN_DNA
C 7 3
C 2 17
T 3 44
…
END_DNA

END_SEQUENCE
```
:::

However the most used file is the binary version `AB1` which could be red out from different programs including [cutepeaks](). It displays the DNA sequence along with its quality peak in each position [Fig. @fig-ab1-trace] 

![A graphic of the `AB1` file format from a Sanger sequencing project displayed in cutepeaks software](imgs-seqan/ab1-trace.png){#fig-ab1-trace}

## Sanger processing workflow

<!-- :::{.callout-tip .column-margin}
A simple one-line to generate a reverse-complement of a linearized FASTA could be done with three commands:

```{.bash}
tail -n 1 seq.fasta | 
    rev | 
    tr "ATCG" "TAGC" 
```

[see here](https://github.com/crazyhottommy/bioinformatics-one-liners#reverse-complement-a-sequence-i-use-that-a-lot-when-i-need-to-design-primers)
::: -->

Despite being one of the oldest sequencing methods, Sanger sequencing has some remarkable strengths that are enable thanks to the automation era. For instance modern sequencers working in parallel can read almost 384 different sequences (of 750bp) in 1h (7Mb in 24h per machine), clearly automation introduces a second advantage and that is the low-price of sequencing small fragments @brown2018. It is important to highlight some limitations as well like the necessity of round-the-clock technical support (this also leverages a little bit from robotic operators though). The error-free Sanger sequencing requires at least 5X of sample coverage, which could be seen as disadvantage too @brown2018.


## The 16S rRNA and its relevance for sequencing



![*Bacillus subtilis* subsp. *subtilis* str. 168 bacterial SSU 16S rRNA](imgs-seqan/16-rRNA.png){#fig-16-rna width=80% .column-margin}